---
permalink: /doc/samtools.html
layout: manpage
title: samtools(1) manual page
package: samtools-1.15
date: 21 February 2022
section_text: Bioinformatics tools
redirect_from: /doc/samtools.1.html
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools &ndash; Utilities for the Sequence Alignment/Map (SAM) format
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
samtools <a href="samtools-addreplacerg.html">addreplacerg</a> -r 'ID:fish' -r 'LB:1334' -r 'SM:alpha' -o output.bam input.bam
<p>
samtools <a href="samtools-ampliconclip.html">ampliconclip</a> -b bed.file input.bam
<p>
samtools <a href="samtools-ampliconstats.html">ampliconstats</a> primers.bed in.bam
<p>
samtools <a href="samtools-bedcov.html">bedcov</a> aln.sorted.bam
<p>
samtools <a href="samtools-calmd.html">calmd</a> in.sorted.bam ref.fasta
<p>
samtools <a href="samtools-cat.html">cat</a> out.bam in1.bam in2.bam in3.bam
<p>
samtools <a href="samtools-collate.html">collate</a> -o aln.name_collated.bam aln.sorted.bam
<p>
samtools <a href="samtools-consensus.html">consensus</a> -o out.fasta in.bam
<p>
samtools <a href="samtools-coverage.html">coverage</a> aln.sorted.bam
<p>
samtools <a href="samtools-depad.html">depad</a> input.bam
<p>
samtools <a href="samtools-depth.html">depth</a> aln.sorted.bam
<p>
samtools <a href="samtools-dict.html">dict</a> -a GRCh38 -s "Homo sapiens" ref.fasta
<p>
samtools <a href="samtools-faidx.html">faidx</a> ref.fasta
<p>
samtools <a href="samtools-fasta.html">fasta</a> input.bam &gt; output.fasta
<p>
samtools <a href="samtools-fastq.html">fastq</a> input.bam &gt; output.fastq
<p>
samtools <a href="samtools-fixmate.html">fixmate</a> in.namesorted.sam out.bam
<p>
samtools <a href="samtools-flags.html">flags</a> PAIRED,UNMAP,MUNMAP
<p>
samtools <a href="samtools-flagstat.html">flagstat</a> aln.sorted.bam
<p>
samtools <a href="samtools-fqidx.html">fqidx</a> ref.fastq
<p>
samtools <a href="samtools-head.html">head</a> in.bam
<p>
samtools <a href="samtools-idxstats.html">idxstats</a> aln.sorted.bam
<p>
samtools <a href="samtools-import.html">import</a> input.fastq &gt; output.bam
<p>
samtools <a href="samtools-index.html">index</a> aln.sorted.bam
<p>
samtools <a href="samtools-markdup.html">markdup</a> in.algnsorted.bam out.bam
<p>
samtools <a href="samtools-merge.html">merge</a> out.bam in1.bam in2.bam in3.bam
<p>
samtools <a href="samtools-mpileup.html">mpileup</a> -C50 -f ref.fasta -r chr3:1,000-2,000 in1.bam in2.bam
<p>
samtools <a href="samtools-phase.html">phase</a> input.bam
<p>
samtools <a href="samtools-quickcheck.html">quickcheck</a> in1.bam in2.cram
<p>
samtools <a href="samtools-reheader.html">reheader</a> in.header.sam in.bam &gt; out.bam
<p>
samtools <a href="samtools-samples.html">samples</a> input.bam
<p>
samtools <a href="samtools-sort.html">sort</a> -T /tmp/aln.sorted -o aln.sorted.bam aln.bam
<p>
samtools <a href="samtools-split.html">split</a> merged.bam
<p>
samtools <a href="samtools-stats.html">stats</a> aln.sorted.bam
<p>
samtools <a href="samtools-targetcut.html">targetcut</a> input.bam
<p>
samtools <a href="samtools-tview.html">tview</a> aln.sorted.bam ref.fasta
<p>
samtools <a href="samtools-view.html">view</a> -bt ref_list.txt -o aln.bam aln.sam.gz
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
Samtools is a set of utilities that manipulate alignments in the SAM
(Sequence Alignment/Map), BAM, and CRAM formats.
It converts between the formats, does sorting, merging and indexing,
and can retrieve reads in any regions swiftly.
<p>
Samtools is designed to work on a stream. It regards an input file `-'
as the standard input (stdin) and an output file `-' as the standard
output (stdout). Several commands can thus be combined with Unix
pipes. Samtools always output warning and error messages to the standard
error output (stderr).
<p>
Samtools is also able to open files on remote FTP or
HTTP(S) servers if the file name starts with `ftp://', `http://', etc.
Samtools checks the current working directory for the index file and
will download the index upon absence. Samtools does not retrieve the
entire alignment file unless it is asked to do so.
<p>
If an index is needed, samtools looks for the index suffix
appended to the filename, and if that isn't found it tries again
without the filename suffix (for example <b>in.bam.bai</b> followed by
<b>in.bai</b>).  However if an index is in a completely different
location or has a different name, both the main data filename and
index filename can be pasted together with <b>##idx##</b>.  For
example <b>/data/in.bam##idx##/indices/in.bam.bai</b> may be used to
explicitly indicate where the data and index files reside.
<p>
<h1 id="COMMANDS"><a href="#COMMANDS">COMMANDS</a></h1>
<p>
Each command has its own man page which can be viewed using
e.g. <b>man samtools-view</b> or with a recent GNU man using
<b>man samtools view</b>.  Below we have a brief summary of syntax
and sub-command description.
<p>
Options common to all sub-commands are documented below in the GLOBAL
COMMAND OPTIONS section.
<p>
<dl><dt><b>view</b></dt><dd><p>
samtools view
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[<em>region</em>...]
<p>
With no options or regions specified, prints all alignments in the specified
input alignment file (in SAM, BAM, or CRAM format) to standard output
in SAM format (with no header by default).
<p>
You may specify one or more space-separated region specifications after the
input filename to restrict output to only those alignments which overlap the
specified region(s). Use of region specifications requires a coordinate-sorted
and indexed input file.
<p>
Options exist to change the output format from SAM to BAM or CRAM, so
this command also acts as a file format conversion utility.
<p>
</dd><dt><b>tview</b></dt><dd><p>
samtools tview
[<b>-p</b>
<em>chr:pos</em>]
[<b>-s</b>
<em>STR</em>]
[<b>-d</b>
<em>display</em>]
&lt;in.sorted.bam&gt;
[ref.fasta]
<p>
Text alignment viewer (based on the ncurses library). In the viewer,
press `?' for help and press `g' to check the alignment start from a
region in the format like `chr10:10,000,000' or `=10,000,000' when
viewing the same reference sequence.
<p>
</dd><dt><b>quickcheck</b></dt><dd><p>
samtools quickcheck
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[ ... ]
<p>
Quickly check that input files appear to be intact. Checks that beginning of the
file contains a valid header (all formats) containing at least one target
sequence and then seeks to the end of the file and checks that an end-of-file
(EOF) is present and intact (BAM only).
<p>
Data in the middle of the file is not read since that would be much more time
consuming, so please note that this command will not detect internal corruption,
but is useful for testing that files are not truncated before performing more
intensive tasks on them.
<p>
This command will exit with a non-zero exit code if any input files don't have a
valid header or are missing an EOF block. Otherwise it will exit successfully
(with a zero exit code).
<p>
</dd><dt><b>head</b></dt><dd><p>
samtools head
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
<p>
Prints the input file's headers and optionally also its first few alignment
records. This command always displays the headers as they are in the file,
never adding an extra @PG header itself.
<p>
</dd><dt><b>index</b></dt><dd><p>
samtools index
[<b>-bc</b>]
[<b>-m</b>
<em>INT</em>]
<em>aln.sam.gz</em>|<em>aln.bam</em>|<em>aln.cram</em>
[<em>out.index</em>]
<p>
Index a coordinate-sorted SAM, BAM or CRAM file for fast random access.
Note for SAM this only works if the file has been BGZF compressed first.
<p>
This index is needed when
<em>region</em>
arguments are used to limit
<b>samtools view</b>
and similar commands to particular regions of interest.
<p>
If an output filename is given, the index file will be written to
<em>out.index</em>.
Otherwise, for a CRAM file
<em>aln.cram</em>,
index file
<em>aln.cram</em><b>.crai</b>
will be created; for a BAM or SAM file
<em>aln.bam</em>,
either
<em>aln.bam</em><b>.bai</b>
or
<em>aln.bam</em><b>.csi</b>
will be created, depending on the index format selected.
<p>
</dd><dt><b>sort</b></dt><dd><p>
samtools sort
[<b>-l</b>
<em>level</em>]
[<b>-m</b>
<em>maxMem</em>]
[<b>-o</b>
<em>out.bam</em>]
[<b>-O</b>
<em>format</em>]
[<b>-n</b>]
[<b>-t</b>
<em>tag</em>]
[<b>-T</b>
<em>tmpprefix</em>]
[<b>-@</b>
<em>threads</em>] [<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>]
<p>
Sort alignments by leftmost coordinates, or by read name when
<b>-n</b>
is used.
An appropriate
<b>@HD-SO</b>
sort order header tag will be added or an existing one updated if necessary.
<p>
The sorted output is written to standard output by default, or to the
specified file
(<em>out.bam</em>)
when
<b>-o</b>
is used.
This command will also create temporary files
<em>tmpprefix</em><b>.</b><em>%d</em><b>.bam</b>
as needed when the entire alignment data cannot fit into memory
(as controlled via the
<b>-m</b>
option).
<p>
Consider using
<b>samtools collate</b>
instead if you need name collated data without a full lexicographical sort.
<p>
</dd><dt><b>collate</b></dt><dd><p>
samtools collate
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em> [<em>&lt;prefix&gt;</em>]
<p>
Shuffles and groups reads together by their names.
A faster alternative to a full query name sort,
<b>collate</b>
ensures that reads of the same name are grouped together in contiguous groups,
but doesn't make any guarantees about the order of read names between groups.
<p>
The output from this command should be suitable for any operation that
requires all reads from the same template to be grouped together.
<p>
</dd><dt><b>idxstats</b></dt><dd><p>
samtools idxstats
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
<p>
Retrieve and print stats in the index file corresponding to the input file.
Before calling idxstats, the input BAM file should be indexed by samtools index.
<p>
If run on a SAM or CRAM file or an unindexed BAM file, this command
will still produce the same summary statistics, but does so by reading
through the entire file.  This is far slower than using the BAM
indices.
<p>
The output is TAB-delimited with each line consisting of reference sequence
name, sequence length, # mapped reads and # unmapped reads. It is written to
stdout.
<p>
</dd><dt><b>flagstat</b></dt><dd><p>
samtools flagstat
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
<p>
Does a full pass through the input file to calculate and print statistics
to stdout.
<p>
Provides counts for each of 13 categories based primarily on bit flags in
the FLAG field. Each category in the output is broken down into QC pass and
QC fail, which is presented as "#PASS + #FAIL" followed by a description of
the category.
<p>
</dd><dt><b>flags</b></dt><dd><p>
samtools flags
<em>INT</em>|<em>STR</em>[,...]
<p>
Convert between textual and numeric flag representation.
<p>
<b>FLAGS:</b>
<table>
<tr><td><b>0x1</b></td><td>PAIRED</td><td>paired-end (or multiple-segment) sequencing technology</td></tr>
<tr><td><b>0x2</b></td><td>PROPER_PAIR</td><td>each segment properly aligned according to the aligner</td></tr>
<tr><td><b>0x4</b></td><td>UNMAP</td><td>segment unmapped</td></tr>
<tr><td><b>0x8</b></td><td>MUNMAP</td><td>next segment in the template unmapped</td></tr>
<tr><td><b>0x10</b></td><td>REVERSE</td><td>SEQ is reverse complemented</td></tr>
<tr><td><b>0x20</b></td><td>MREVERSE</td><td>SEQ of the next segment in the template is reverse complemented</td></tr>
<tr><td><b>0x40</b></td><td>READ1</td><td>the first segment in the template</td></tr>
<tr><td><b>0x80</b></td><td>READ2</td><td>the last segment in the template</td></tr>
<tr><td><b>0x100</b></td><td>SECONDARY</td><td>secondary alignment</td></tr>
<tr><td><b>0x200</b></td><td>QCFAIL</td><td>not passing quality controls</td></tr>
<tr><td><b>0x400</b></td><td>DUP</td><td>PCR or optical duplicate</td></tr>
<tr><td><b>0x800</b></td><td>SUPPLEMENTARY</td><td>supplementary alignment</td></tr>
</table>
<p>
</dd><dt><b>stats</b></dt><dd><p>
samtools stats
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[<em>region</em>...]
<p>
samtools stats collects statistics from BAM files and outputs in a text format.
The output can be visualized graphically using plot-bamstats.
<p>
<p>
</dd><dt><b>bedcov</b></dt><dd><p>
samtools bedcov
[<em>options</em>]
<em>region.bed</em> <em>in1.sam</em>|<em>in1.bam</em>|<em>in1.cram</em>[...]
<p>
Reports the total read base count (i.e. the sum of per base read depths)
for each genomic region specified in the supplied BED file. The regions
are output as they appear in the BED file and are 0-based.
Counts for each alignment file supplied are reported in separate columns.
<p>
</dd><dt><b>depth</b></dt><dd><p>
samtools depth
[<em>options</em>]
[<em>in1.sam</em>|<em>in1.bam</em>|<em>in1.cram</em> [<em>in2.sam</em>|<em>in2.bam</em>|<em>in2.cram</em>] [...]]
<p>
Computes the read depth at each position or region.
<p>
</dd><dt><b>ampliconstats</b></dt><dd><p>
samtools ampliconstats
[<em>options</em>]
<em>primers.bed</em>
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>[...]
<p>
samtools ampliconstats collects statistics from one or more input
alignment files and produces tables in text format.  The output can be
visualized graphically using plot-ampliconstats.
<p>
The alignment files should have previously been clipped of primer
sequence, for example by <b>samtools ampliconclip</b> and the sites of
these primers should be specified as a bed file in the arguments.
<p>
</dd><dt><b>mpileup</b></dt><dd><p>
samtools mpileup
[<b>-EB</b>]
[<b>-C</b>
<em>capQcoef</em>]
[<b>-r</b>
<em>reg</em>]
[<b>-f</b>
<em>in.fa</em>]
[<b>-l</b>
<em>list</em>]
[<b>-Q</b>
<em>minBaseQ</em>]
[<b>-q</b>
<em>minMapQ</em>]
<em>in.bam</em>
[<em>in2.bam</em>
[<em>...</em>]]
<p>
Generate textual pileup for one or multiple BAM files.  For VCF and
BCF output, please use the
<b>bcftools mpileup</b>
command instead.
Alignment records are grouped by sample (SM) identifiers in @RG header lines.
If sample identifiers are absent, each input file is regarded as one sample.
<p>
See the samtools-mpileup man page for a description of the pileup format and options.
<p>
</dd><dt><b>consensus</b></dt><dd><p>
samtools consensus
[<b>options</b>]
<em>in.bam</em>
<p>
Generate consensus from a SAM, BAM or CRAM file based on the contents
of the alignment records.  The consensus is written either as FASTA, 
FASTQ, or a pileup oriented format.
<p>
The default output for FASTA and FASTQ formats include one base per
non-gap consensus.  Hence insertions with respect to the aligned
reference will be included and deletions removed.  This behaviour can
be adjusted.
<p>
Two consensus calling algorithms are offered.  The default computes a
heterozygous consensus in a Bayesian manner, derived from the "Gap5"
consensus algorithm.  A simpler base frequency counting method is also
available.
<p>
<p>
</dd><dt><b>coverage</b></dt><dd><p>
samtools coverage
[<em>options</em>]
[<em>in1.sam</em>|<em>in1.bam</em>|<em>in1.cram</em> [<em>in2.sam</em>|<em>in2.bam</em>|<em>in2.cram</em>] [...]]
<p>
Produces a histogram or table of coverage per chromosome.
<p>
</dd><dt><b>merge</b></dt><dd><p>
samtools merge
[<b>-nur1f</b>]
[<b>-h</b>
<em>inh.sam</em>]
[<b>-t</b>
<em>tag</em>]
[<b>-R</b>
<em>reg</em>]
[<b>-b</b>
<em>list</em>] <em>out.bam</em> <em>in1.bam</em> [<em>in2.bam</em> <em>in3.bam</em> ... <em>inN.bam</em>]
<p>
Merge multiple sorted alignment files, producing a single sorted output file
that contains all the input records and maintains the existing sort order.
<p>
If
<b>-h</b>
is specified the @SQ headers of input files will be merged into the specified header, otherwise they will be merged
into a composite header created from the input headers.
If the @SQ headers differ in order this may require the output file to be re-sorted after merge.
<p>
The ordering of the records in the input files must match the usage of the
<b>-n</b> and <b>-t</b> command-line options.  If they do not, the output
order will be undefined.  See
<b>sort</b>
for information about record ordering.
<p>
</dd><dt><b>split</b></dt><dd><p>
samtools split
[<em>options</em>]
<em>merged.sam</em>|<em>merged.bam</em>|<em>merged.cram</em>
<p>
Splits a file by read group, producing one or more output files
matching a common prefix (by default based on the input filename)
each containing one read-group.
<p>
</dd><dt><b>cat</b></dt><dd><p>
samtools cat
[<b>-b</b>
<em>list</em>]
[<b>-h</b>
<em>header.sam</em>]
[<b>-o</b>
<em>out.bam</em>] <em>in1.bam</em> <em>in2.bam</em> [ ... ]
<p>
Concatenate BAMs or CRAMs. Although this works on either BAM or CRAM,
all input files must be the same format as each other. The sequence
dictionary of each input file must be identical, although this command
does not check this. This command uses a similar trick to
<b>reheader</b>
which enables fast BAM concatenation.
<p>
</dd><dt><b>import</b></dt><dd><p>
samtools import
[<em>options</em>]
<em>in.fastq</em> [ ... ]
<p>
Converts one or more FASTQ files to unaligned SAM, BAM or CRAM.  These
formats offer a richer capability of tracking sample meta-data via the
SAM header and per-read meta-data via the auxiliary tags.  The
<b>fastq</b>
command may be used to reverse this conversion.
<p>
</dd><dt><b>fastq/a</b></dt><dd><p>
samtools fastq
[<em>options</em>]
<em>in.bam</em>
<br>
samtools fasta
[<em>options</em>]
<em>in.bam</em>
<p>
Converts a BAM or CRAM into either FASTQ or FASTA format depending on the
command invoked. The files will be automatically compressed if the
file names have a .gz or .bgzf extension.
<p>
The input to this program must be collated by name.
Use
<b>samtools collate</b>
or
<b>samtools sort -n</b>
to ensure this.
<p>
</dd><dt><b>faidx</b></dt><dd><p>
samtools faidx &lt;ref.fasta&gt; [region1 [...]]
<p>
Index reference sequence in the FASTA format or extract subsequence from
indexed reference sequence. If no region is specified,
<b>faidx</b>
will index the file and create
<em>&lt;ref.fasta&gt;.fai</em>
on the disk. If regions are specified, the subsequences will be
retrieved and printed to stdout in the FASTA format.
<p>
The input file can be compressed in the
<b>BGZF</b>
format.
<p>
FASTQ files can be read and indexed by this command.  Without using
<b>--fastq</b>
any extracted subsequence will be in FASTA format.
<p>
</dd><dt><b>fqidx</b></dt><dd><p>
samtools fqidx &lt;ref.fastq&gt; [region1 [...]]
<p>
Index reference sequence in the FASTQ format or extract subsequence from
indexed reference sequence. If no region is specified,
<b>fqidx</b>
will index the file and create
<em>&lt;ref.fastq&gt;.fai</em>
on the disk. If regions are specified, the subsequences will be
retrieved and printed to stdout in the FASTQ format.
<p>
The input file can be compressed in the
<b>BGZF</b>
format.
<p>
<b>samtools fqidx</b>
should only be used on fastq files with a small number of entries.
Trying to use it on a file containing millions of short sequencing reads
will produce an index that is almost as big as the original file, and
searches using the index will be very slow and use a lot of memory.
<p>
</dd><dt><b>dict</b></dt><dd><p>
samtools dict
<em>ref.fasta</em>|<em>ref.fasta.gz</em>
<p>
Create a sequence dictionary file from a fasta file.
<p>
</dd><dt><b>calmd</b></dt><dd><p>
samtools calmd
[<b>-Eeubr</b>]
[<b>-C</b>
<em>capQcoef</em>] <em>aln.bam</em> <em>ref.fasta</em>
<p>
Generate the MD tag. If the MD tag is already present, this command will
give a warning if the MD tag generated is different from the existing
tag. Output SAM by default.
<p>
Calmd can also read and write CRAM files although in most cases it is
pointless as CRAM recalculates MD and NM tags on the fly.  The one
exception to this case is where both input and output CRAM files
have been / are being created with the <em>no_ref</em> option.
<p>
</dd><dt><b>fixmate</b></dt><dd><p>
samtools fixmate
[<b>-rpcm</b>]
[<b>-O</b>
<em>format</em>]
<em>in.nameSrt.bam out.bam</em>
<p>
Fill in mate coordinates, ISIZE and mate related flags from a
name-sorted alignment.
<p>
</dd><dt><b>markdup</b></dt><dd><p>
samtools markdup
[<b>-l</b>
<em>length</em>]
[<b>-r</b>]
[<b>-s</b>]
[<b>-T</b>]
[<b>-S</b>]
<em>in.algsort.bam out.bam</em>
<p>
Mark duplicate alignments from a coordinate sorted file that
has been run through <b>samtools fixmate</b> with the <b>-m</b> option.  This program
relies on the MC and ms tags that fixmate provides.
<p>
</dd><dt><b>rmdup</b></dt><dd><p>
samtools rmdup [-sS] &lt;input.srt.bam&gt; &lt;out.bam&gt;
<p>
<b>This command is obsolete. Use markdup instead.</b>
<p>
</dd><dt><b>addreplacerg</b></dt><dd><p>
samtools addreplacerg
[<b>-r</b>
<em>rg-line</em>
|
<b>-R</b>
<em>rg-ID</em>]
[<b>-m</b>
<em>mode</em>]
[<b>-l</b>
<em>level</em>]
[<b>-o</b>
<em>out.bam</em>]
<em>in.bam</em>
<p>
Adds or replaces read group tags in a file.
<p>
</dd><dt><b>reheader</b></dt><dd><p>
samtools reheader
[<b>-iP</b>]
<em>in.header.sam in.bam</em>
<p>
Replace the header in
<em>in.bam</em>
with the header in
<em>in.header.sam</em>.
This command is much faster than replacing the header with a
BAM&rarr;SAM&rarr;BAM conversion.
<p>
By default this command outputs the BAM or CRAM file to standard
output (stdout), but for CRAM format files it has the option to
perform an in-place edit, both reading and writing to the same file.
No validity checking is performed on the header, nor that it is suitable
to use with the sequence data itself.
<p>
</dd><dt><b>targetcut</b></dt><dd><p>
samtools targetcut
[<b>-Q</b>
<em>minBaseQ</em>]
[<b>-i</b>
<em>inPenalty</em>]
[<b>-0</b>
<em>em0</em>]
[<b>-1</b>
<em>em1</em>]
[<b>-2</b>
<em>em2</em>]
[<b>-f</b>
<em>ref</em>] <em>in.bam</em>
<p>
This command identifies target regions by examining the continuity of read depth, computes
haploid consensus sequences of targets and outputs a SAM with each sequence corresponding
to a target. When option
<b>-f</b>
is in use, BAQ will be applied. This command is
<b>only</b>
designed for cutting fosmid clones from fosmid pool sequencing [Ref. Kitzman et al. (2010)].
<p>
</dd><dt><b>phase</b></dt><dd><p>
samtools phase
[<b>-AF</b>]
[<b>-k</b>
<em>len</em>]
[<b>-b</b>
<em>prefix</em>]
[<b>-q</b>
<em>minLOD</em>]
[<b>-Q</b>
<em>minBaseQ</em>] <em>in.bam</em>
<p>
Call and phase heterozygous SNPs.
<p>
</dd><dt><b>depad</b></dt><dd><p>
samtools depad
[<b>-SsCu1</b>]
[<b>-T</b>
<em>ref.fa</em>]
[<b>-o</b>
<em>output</em>] <em>in.bam</em>
<p>
Converts a BAM aligned against a padded reference to a BAM aligned
against the depadded reference.  The padded reference may contain
verbatim "*" bases in it, but "*" bases are also counted in the
reference numbering.  This means that a sequence base-call aligned
against a reference "*" is considered to be a cigar match ("M" or "X")
operator (if the base-call is "A", "C", "G" or "T").  After depadding
the reference "*" bases are deleted and such aligned sequence
base-calls become insertions.  Similarly transformations apply for
deletions and padding cigar operations.
<p>
</dd><dt><b>ampliconclip</b></dt><dd><p>
samtools ampliconclip
[<b>-o</b>
<em>out.file</em>]
[<b>-f</b>
<em>stat.file</em>]
[<b>--soft-clip</b>]
[<b>--hard-clip</b>]
[<b>--both-ends</b>]
[<b>--strand</b>]
[<b>--clipped</b>]
[<b>--fail</b>]
[<b>--no-PG</b>]
<b>-b</b>
<em>bed.file in.file</em>
<p>
Clip reads in a SAM compatible file based on data from a BED file.
<p>
</dd><dt><b>samples</b></dt><dd><p>
samtools samples
[<b>-o</b>
<em>out.file</em>]
[<b>-i</b>]
[<b>-T</b>
<em>TAG</em>]
[<b>-f</b>
<em>refs.fasta</em>]
[<b>-F</b>
<em>refs_list</em>]
[<b>-X</b>]
<p>
Prints the samples from alignment files
<p>
</dd></dl>
<h1 id="SAMTOOLS_OPTIONS"><a href="#SAMTOOLS_OPTIONS">SAMTOOLS OPTIONS</a></h1>
These are options that are passed after the <b>samtools</b> command,
before any sub-command is specified.
</pre>
<dl><dt><b>help</b>, <b>--help</b></dt><dd><p>
Display a brief usage message listing the samtools commands available.
If the name of a command is also given, e.g.,
<b>samtools help view</b>,
the detailed usage message for that particular command is displayed.
<p>
</dd><dt><b>--version</b></dt><dd><p>
Display the version numbers and copyright information for samtools and
the important libraries used by samtools.
<p>
</dd><dt><b>--version-only</b></dt><dd><p>
Display the full samtools version number in a machine-readable format.
</dd></dl><p>
<h1 id="GLOBAL_COMMAND_OPTIONS"><a href="#GLOBAL_COMMAND_OPTIONS">GLOBAL COMMAND OPTIONS</a></h1>
<p>
Several long-options are shared between multiple samtools sub-commands:
<b>--input-fmt</b>, <b>--input-fmt-option</b>, <b>--output-fmt</b>,
<b>--output-fmt-option</b>, <b>--reference</b>, <b>--write-index</b>,
and <b>--verbosity</b>.
The input format is typically auto-detected so specifying the format
is usually unnecessary and the option is included for completeness.
Note that not all subcommands have all options.  Consult the subcommand
help for more details.
<p>
Format strings recognised are "sam", "sam.gz", "bam" and "cram".  They may be
followed by a comma separated list of options as <em>key</em> or
<em>key</em>=<em>value</em>. See below for examples.
<p>
The <b>fmt-option</b> arguments accept either a single <em>option</em> or
<em>option</em>=<em>value</em>.  Note that some options only work on some
file formats and only on read or write streams.  If value is
unspecified for a boolean option, the value is assumed to be 1.  The
valid options are as follows.
<dl><dt><b>level=</b><em>INT</em></dt><dd><p>
Output only. Specifies the compression level from 1 to 9, or 0 for
uncompressed.  If the output format is SAM, this also enables BGZF
compression, otherwise SAM defaults to uncompressed.
</dd><dt><b>nthreads=</b><em>INT</em></dt><dd><p>
Specifies the number of threads to use during encoding and/or
decoding.  For BAM this will be encoding only.  In CRAM the threads
are dynamically shared between encoder and decoder.
</dd><dt><b>filter=</b><em>STRING</em></dt><dd><p>
Apply filter STRING to all incoming records, rejecting any that do not
satisfy the expression.  See the FILTER EXPRESSIONS section below for
specifics.
</dd><dt><b>reference=</b><em>fasta_file</em></dt><dd><p>
Specifies a FASTA reference file for use in CRAM encoding or decoding.
It usually is not required for decoding except in the situation of the
MD5 not being obtainable via the REF_PATH or REF_CACHE environment variables.
</dd><dt><b>decode_md=</b><em>0|1</em></dt><dd><p>
CRAM input only; defaults to 1 (on).  CRAM does not typically store
MD and NM tags, preferring to generate them on the fly.  When this
option is 0 missing MD, NM tags will not be generated.  It can be
particularly useful when combined with a file encoded using store_md=1
and store_nm=1.
</dd><dt><b>store_md=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  CRAM normally only stores MD
tags when the reference is unknown and lets the decoder generate these
values on-the-fly (see decode_md).
</dd><dt><b>store_nm=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  CRAM normally only stores NM
tags when the reference is unknown and lets the decoder generate these
values on-the-fly (see decode_md).
</dd><dt><b>ignore_md5=</b><em>0|1</em></dt><dd><p>
CRAM input only; defaults to 0 (off).  When enabled, md5 checksum
errors on the reference sequence and block checksum errors within CRAM
are ignored.  Use of this option is strongly discouraged.
</dd><dt><b>required_fields=</b><em>bit-field</em></dt><dd><p>
CRAM input only; specifies which SAM columns need to be populated.
By default all fields are used.  Limiting the decode to specific
columns can have significant performance gains.  The bit-field is a
numerical value constructed from the following table.
<table>
<tr><td><b>0x1</b></td><td>SAM_QNAME</td></tr>
<tr><td><b>0x2</b></td><td>SAM_FLAG</td></tr>
<tr><td><b>0x4</b></td><td>SAM_RNAME</td></tr>
<tr><td><b>0x8</b></td><td>SAM_POS</td></tr>
<tr><td><b>0x10</b></td><td>SAM_MAPQ</td></tr>
<tr><td><b>0x20</b></td><td>SAM_CIGAR</td></tr>
<tr><td><b>0x40</b></td><td>SAM_RNEXT</td></tr>
<tr><td><b>0x80</b></td><td>SAM_PNEXT</td></tr>
<tr><td><b>0x100</b></td><td>SAM_TLEN</td></tr>
<tr><td><b>0x200</b></td><td>SAM_SEQ</td></tr>
<tr><td><b>0x400</b></td><td>SAM_QUAL</td></tr>
<tr><td><b>0x800</b></td><td>SAM_AUX</td></tr>
<tr><td><b>0x1000</b></td><td>SAM_RGAUX</td></tr>
</table>
</dd><dt><b>name_prefix=</b><em>string</em></dt><dd><p>
CRAM input only; defaults to output filename.  Any sequences with
auto-generated read names will use <em>string</em> as the name prefix.
</dd><dt><b>multi_seq_per_slice=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  By default CRAM generates one
container per reference sequence, except in the case of many small
references (such as a fragmented assembly).
</dd><dt><b>version=</b><em>major.minor</em></dt><dd><p>
CRAM output only.  Specifies the CRAM version number.  Acceptable
values are "2.1" and "3.0".
</dd><dt><b>seqs_per_slice=</b><em>INT</em></dt><dd><p>
CRAM output only; defaults to 10000.
</dd><dt><b>slices_per_container=</b><em>INT</em></dt><dd><p>
CRAM output only; defaults to 1.  The effect of having multiple slices
per container is to share the compression header block between
multiple slices.  This is unlikely to have any significant impact
unless the number of sequences per slice is reduced.  (Together these
two options control the granularity of random access.)
</dd><dt><b>embed_ref=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  If 1, this will store portions
of the reference sequence in each slice, permitting decode without
having requiring an external copy of the reference sequence.
</dd><dt><b>no_ref=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  If 1, sequences will be stored
verbatim with no reference encoding.  This can be useful if no
reference is available for the file.
</dd><dt><b>use_bzip2=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  Permits use of bzip2 in CRAM
block compression.
</dd><dt><b>use_lzma=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  Permits use of lzma in CRAM
block compression.
</dd><dt><b>lossy_names=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  If 1, templates with all
members within the same CRAM slice will have their read names
removed.  New names will be automatically generated during decoding.
Also see the <b>name_prefix</b> option.
</dd></dl>
<p>
For example:
<pre>
samtools view --input-fmt-option decode_md=0
    --output-fmt cram,version=3.0 --output-fmt-option embed_ref
    --output-fmt-option seqs_per_slice=2000 -o foo.cram foo.bam
</pre>
<p>
The <b>--write-index</b> option enables automatic index creation while
writing out BAM, CRAM or bgzf SAM files.  Note to get compressed SAM
as the output format you need to manually request a compression level,
otherwise all SAM files are uncompressed.  By default SAM and BAM will
use CSI indices while CRAM will use CRAI indices.  If you need to
create BAI indices note that it is possible to specify the name of
the index being written to, and hence the format, by using the
<b>filename##idx##indexname</b> notation.
<p>
For example: to convert a BAM to a compressed SAM with CSI indexing:
<pre>
samtools view -h -O sam,level=6 --write-index in.bam -o out.sam.gz
</pre>
<p>
To convert a SAM to a compressed BAM using BAI indexing:
<pre>
samtools view --write-index in.sam -o out.bam##idx##out.bam.bai
</pre>
<p>
The <b>--verbosity </b><em>INT</em> option sets the verbosity level for samtools
and HTSlib.  The default is 3 (HTS_LOG_WARNING); 2 reduces warning messages
and 0 or 1 also reduces some error messages, while values greater than 3
produce increasing numbers of additional warnings and logging messages.
<p>
<p>
<h1 id="REFERENCE_SEQUENCES"><a href="#REFERENCE_SEQUENCES">REFERENCE SEQUENCES</a></h1>
<p>
The CRAM format requires use of a reference sequence for both reading
and writing.
<p>
When reading a CRAM the <b>@SQ</b> headers are interrogated to identify
the reference sequence MD5sum (<b>M5:</b> tag) and the local reference
sequence filename (<b>UR:</b> tag).  Note that <em>http://</em> and
<em>ftp://</em> based URLs in the UR: field are not used, but local fasta
filenames (with or without <em>file://</em>) can be used.
<p>
To create a CRAM the <b>@SQ</b> headers will also be read to identify
the reference sequences, but M5: and UR: tags may not be present. In
this case the <b>-T</b> and <b>-t</b> options of samtools view may be
used to specify the fasta or fasta.fai filenames respectively
(provided the .fasta.fai file is also backed up by a .fasta file).
<p>
The search order to obtain a reference is:
<p>
Use any local file specified by the command line options (eg -T).
<p>
Look for MD5 via REF_CACHE environment variable.
<p>
Look for MD5 in each element of the REF_PATH environment variable.
<p>
Look for a local file listed in the UR: header tag.
<p>
<p>
<h1 id="FILTER_EXPRESSIONS"><a href="#FILTER_EXPRESSIONS">FILTER EXPRESSIONS</a></h1>
<p>
Filter expressions are used as an on-the-fly checking of incoming SAM,
BAM or CRAM records, discarding records that do not match the
specified expression.
<p>
The language used is primarily C style, but with a few differences in
the precedence rules for bit operators and the inclusion of regular
expression matching.
<p>
The operator precedence, from strongest binding to weakest, is:
<p>
<table>
<tr><td>Grouping</td><td><b>(, )</b></td><td>E.g. "(1+2)*3"</td></tr>
<tr><td>Values:</td><td><b>literals, vars</b></td><td>Numbers, strings and variables</td></tr>
<tr><td>Unary ops:</td><td><b>+, -, !, ~</b></td><td>E.g. -10 +10, !10 (not), ~5 (bit not)</td></tr>
<tr><td>Math ops:</td><td><b>*, /, %</b></td><td>Multiply, division and (integer) modulo</td></tr>
<tr><td>Math ops:</td><td><b>+, -</b></td><td>Addition / subtraction</td></tr>
<tr><td>Bit-wise:</td><td><b>&amp;</b></td><td>Integer AND</td></tr>
<tr><td>Bit-wise</td><td><b>^</b></td><td>Integer XOR</td></tr>
<tr><td>Bit-wise</td><td><b>|</b></td><td>Integer OR</td></tr>
<tr><td>Conditionals:</td><td><b>&gt;, &gt;=, &lt;, &lt;=</b></td></tr>
<tr><td>Equality:</td><td><b>==, !=, =~, !~</b></td><td>=~ and !~ match regular expressions</td></tr>
<tr><td>Boolean:</td><td><b>&amp;&amp;, ||</b></td><td>Logical AND / OR</td></tr>
</table>
<p>
Expressions are computed using floating point mathematics, so "10 / 4"
evaluates to 2.5 rather than 2.  They may be written as integers in
decimal or "0x" plus hexadecimal, and floating point with or without
exponents.However operations that require integers first do an
implicit type conversion, so "7.9 % 5" is 2 and "7.9 &amp; 4.1" is
equivalent to "7 &amp; 4", which is 4.  Strings are always specified using
double quotes.  To get a double quote in a string, use backslash.
Similarly a double backslash is used to get a literal backslash.  For
example <b>ab\\"c\\\\d</b> is the string <b>ab"c\\d</b>.
<p>
Comparison operators are evaluated as a match being 1 and a mismatch
being 0, thus "(2 &gt; 1) + (3 &lt; 5)" evaluates as 2.
<p>
The variables are where the file format specifics are accessed from
the expression.  The variables correspond to SAM fields, for example
to find paired alignments with high mapping quality and a very large
insert size, we may use the expression "<b>mapq &gt;= 30 &amp;&amp; (tlen &gt;= 100000 || tlen &lt;= -100000)</b>".
Valid variable names and their data types are:
<p>
<table>
<tr><td><b>endpos</b></td><td>int</td><td>Alignment end position (1-based)</td></tr>
<tr><td><b>flag</b></td><td>int</td><td>Combined FLAG field</td></tr>
<tr><td><b>flag.paired</b></td><td>int</td><td>Single bit, 0 or 1</td></tr>
<tr><td><b>flag.proper_pair</b></td><td>int</td><td>Single bit, 0 or 2</td></tr>
<tr><td><b>flag.unmap</b></td><td>int</td><td>Single bit, 0 or 4</td></tr>
<tr><td><b>flag.munmap</b></td><td>int</td><td>Single bit, 0 or 8</td></tr>
<tr><td><b>flag.reverse</b></td><td>int</td><td>Single bit, 0 or 16</td></tr>
<tr><td><b>flag.mreverse</b></td><td>int</td><td>Single bit, 0 or 32</td></tr>
<tr><td><b>flag.read1</b></td><td>int</td><td>Single bit, 0 or 64</td></tr>
<tr><td><b>flag.read2</b></td><td>int</td><td>Single bit, 0 or 128</td></tr>
<tr><td><b>flag.secondary</b></td><td>int</td><td>Single bit, 0 or 256</td></tr>
<tr><td><b>flag.qcfail</b></td><td>int</td><td>Single bit, 0 or 512</td></tr>
<tr><td><b>flag.dup</b></td><td>int</td><td>Single bit, 0 or 1024</td></tr>
<tr><td><b>flag.supplementary</b></td><td>int</td><td>Single bit, 0 or 2048</td></tr>
<tr><td><b>library</b></td><td>string</td><td>Library (LB header via RG)</td></tr>
<tr><td><b>mapq</b></td><td>int</td><td>Mapping quality</td></tr>
<tr><td><b>mpos</b></td><td>int</td><td>Synonym for pnext</td></tr>
<tr><td><b>mrefid</b></td><td>int</td><td>Mate reference number (0 based)</td></tr>
<tr><td><b>mrname</b></td><td>string</td><td>Synonym for rnext</td></tr>
<tr><td><b>ncigar</b></td><td>int</td><td>Number of cigar operations</td></tr>
<tr><td><b>pnext</b></td><td>int</td><td>Mate's alignment position (1-based)</td></tr>
<tr><td><b>pos</b></td><td>int</td><td>Alignment position (1-based)</td></tr>
<tr><td><b>qlen</b></td><td>int</td><td>Alignment length: no. query bases</td></tr>
<tr><td><b>qname</b></td><td>string</td><td>Query name</td></tr>
<tr><td><b>qual</b></td><td>string</td><td>Quality values (raw, 0 based)</td></tr>
<tr><td><b>refid</b></td><td>int</td><td>Integer reference number (0 based)</td></tr>
<tr><td><b>rlen</b></td><td>int</td><td>Alignment length: no. reference bases</td></tr>
<tr><td><b>rname</b></td><td>string</td><td>Reference name</td></tr>
<tr><td><b>rnext</b></td><td>string</td><td>Mate's reference name</td></tr>
<tr><td><b>seq</b></td><td>string</td><td>Sequence</td></tr>
<tr><td><b>tlen</b></td><td>int</td><td>Template length (insert size)</td></tr>
<tr><td><b>[XX]</b></td><td>int / string</td><td>XX tag value</td></tr>
</table>
<p>
Flags are returned either as the whole flag value or by checking for a
single bit.  Hence the filter expression <b>flag.dup</b> is
equivalent to <b>flag &amp; 1024</b>.
<p>
"qlen" and "rlen" are measured using the CIGAR string to count the
number of query (sequence) and reference bases consumed.  Note "qlen"
may not exactly match the length of the "seq" field if the sequence is
"*".
<p>
"endpos" is the (1-based inclusive) position of the rightmost mapped base
of the read, as measured using the CIGAR string, and for mapped reads
is equivalent to "pos+rlen-1". For unmapped reads, it is the same as
"pos".
<p>
Reference names may be matched either by their string forms ("rname"
and "mrname") or as the Nth <b>@SQ</b> line (counting from zero) as
stored in BAM using "tid" and "mtid" respectively.
<p>
Auxiliary tags are described in square brackets and these expand to
either integer or string as defined by the tag itself (<b>XX:Z:</b><em>string</em> or
<b>XX:i:</b><em>int</em>).  For example <b>[NM]&gt;=10</b> can be used to look
for alignments with many mismatches and <b>[RG]=~"grp[ABC]-"</b> will
match the read-group string.
<p>
If no comparison is used with an auxiliary tag it is taken simply to
be a test for the existence of that tag.  So "[NM]" will return any
record containing an NM tag, even if that tag is zero (<b>NM:i:0</b>).
<p>
If you need to check specifically for a non-zero value then use <b>[NM]
&amp;&amp; [NM]!=0</b>.
<p>
Some simple functions are available to operate on strings.  These
treat the strings as arrays of bytes, permitting their length,
minimum, maximum and average values to be computed.
<p>
<table>
<tr><td><b>length</b></td><td>Length of the string (excluding nul char)</td></tr>
<tr><td><b>min</b></td><td>Minimum byte value in the string</td></tr>
<tr><td><b>max</b></td><td>Maximum byte value in the string</td></tr>
<tr><td><b>avg</b></td><td>Average byte value in the string</td></tr>
</table>
<p>
Note that "avg" is a floating point value and it may be NAN for empty
strings.  This means that "avg(qual)" does not produce an error for
records that have both seq and qual of "*".  This value will fail any
conditional checks, so e.g. "avg(qual) &gt; 20" works and will not report
these records.
<p>
<p>
<h1 id="ENVIRONMENT_VARIABLES"><a href="#ENVIRONMENT_VARIABLES">ENVIRONMENT VARIABLES</a></h1>
<p>
<dl><dt><b>HTS_PATH</b></dt><dd><p>
A colon-separated list of directories in which to search for HTSlib plugins.
If $HTS_PATH starts or ends with a colon or contains a double colon (<b>::</b>),
the built-in list of directories is searched at that point in the search.
<p>
If no HTS_PATH variable is defined, the built-in list of directories
specified when HTSlib was built is used, which typically includes
<b>/usr/local/libexec/htslib</b> and similar directories.
<p>
</dd><dt><b>REF_PATH</b></dt><dd><p>
A colon separated (semi-colon on Windows) list of locations in which
to look for sequences identified by their MD5sums.  This can be either
a list of directories or URLs. Note that if a URL is included then the
colon in http:// and ftp:// and the optional port number will be
treated as part of the URL and not a PATH field separator.
For URLs, the text <b>%s</b> will be replaced by the MD5sum being
read.
<p>
If no REF_PATH has been specified it will default to
<b>http://www.ebi.ac.uk/ena/cram/md5/%s</b> and if REF_CACHE is also unset,
it will be set to <b>$XDG_CACHE_HOME/hts-ref/%2s/%2s/%s</b>.
If <b>$XDG_CACHE_HOME</b> is unset, <b>$HOME/.cache</b> (or a local system
temporary directory if no home directory is found) will be used similarly.
<p>
</dd><dt><b>REF_CACHE</b></dt><dd><p>
This can be defined to a single location housing a local cache of
references.  Upon downloading a reference it will be stored in the
location pointed to by REF_CACHE.  REF_CACHE will be searched
before attempting to load via the REF_PATH search list.  If no
REF_PATH is defined, both REF_PATH and REF_CACHE will be automatically
set (see above), but if REF_PATH is defined and REF_CACHE not then no
local cache is used.
<p>
To avoid many files being stored in the same directory, REF_CACHE may
be defined as a pattern using <b>%</b><em>num</em><b>s</b> to consume <em>num</em>
characters of the MD5sum and <b>%s</b> to consume all remaining characters.
If REF_CACHE lacks <b>%s</b> then it will get an implicit <b>/%s</b> appended.
<p>
To aid population of the REF_CACHE directory a script
<b>misc/seq_cache_populate.pl</b> is provided in the Samtools
distribution. This takes a fasta file or a directory of fasta files
and generates the MD5sum named files.
<p>
For example if you use <b>seq_cache_populate -subdirs 2 -root
/local/ref_cache</b> to create 2 nested subdirectories (the default),
each consuming 2 characters of the MD5sum, then REF_CACHE must be set
to <b>/local/ref_cache/%2s/%2s/%s</b>.
</dd></dl><p>
<h1 id="EXAMPLES"><a href="#EXAMPLES">EXAMPLES</a></h1>
<ul><li>Import SAM to BAM when
<b>@SQ</b>
lines are present in the header:
<pre>
samtools view -b aln.sam &gt; aln.bam
</pre>
If
<b>@SQ</b>
lines are absent:
<pre>
samtools faidx ref.fa
samtools view -bt ref.fa.fai aln.sam &gt; aln.bam
</pre>
where
<em>ref.fa.fai</em>
is generated automatically by the
<b>faidx</b>
command.
<p>
</li>
<li>Convert a BAM file to a CRAM file using a local reference sequence.
<pre>
samtools view -C -T ref.fa aln.bam &gt; aln.cram
</pre>
<p>
</li></ul>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Heng Li from the Sanger Institute wrote the original C version of
samtools.  Bob Handsaker from the Broad Institute implemented the BGZF
library.  Petr Danecek and Heng Li wrote the VCF/BCF implementation.
James Bonfield from the Sanger Institute developed the CRAM
implementation.  Other large code contributions have been made by
John Marshall, Rob Davies, Martin Pollard, Andrew Whitwham, Valeriu Ohan
(all while primarily at the Sanger Institute), with numerous other
smaller but valuable contributions.  See the per-command manual pages
for further authorship.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<a href="samtools-addreplacerg.html"><em>samtools-addreplacerg</em></a> (1),
<a href="samtools-ampliconclip.html"><em>samtools-ampliconclip</em></a> (1),
<a href="samtools-ampliconstats.html"><em>samtools-ampliconstats</em></a> (1),
<a href="samtools-bedcov.html"><em>samtools-bedcov</em></a> (1),
<a href="samtools-calmd.html"><em>samtools-calmd</em></a> (1),
<a href="samtools-cat.html"><em>samtools-cat</em></a> (1),
<a href="samtools-collate.html"><em>samtools-collate</em></a> (1),
<a href="samtools-consensus.html"><em>samtools-consensus</em></a> (1),
<a href="samtools-coverage.html"><em>samtools-coverage</em></a> (1),
<a href="samtools-depad.html"><em>samtools-depad</em></a> (1),
<a href="samtools-depth.html"><em>samtools-depth</em></a> (1),
<a href="samtools-dict.html"><em>samtools-dict</em></a> (1),
<a href="samtools-faidx.html"><em>samtools-faidx</em></a> (1),
<a href="samtools-fasta.html"><em>samtools-fasta</em></a> (1),
<a href="samtools-fastq.html"><em>samtools-fastq</em></a> (1),
<a href="samtools-fixmate.html"><em>samtools-fixmate</em></a> (1),
<a href="samtools-flags.html"><em>samtools-flags</em></a> (1),
<a href="samtools-flagstat.html"><em>samtools-flagstat</em></a> (1),
<a href="samtools-fqidx.html"><em>samtools-fqidx</em></a> (1),
<a href="samtools-head.html"><em>samtools-head</em></a> (1),
<a href="samtools-idxstats.html"><em>samtools-idxstats</em></a> (1),
<a href="samtools-import.html"><em>samtools-import</em></a> (1),
<a href="samtools-index.html"><em>samtools-index</em></a> (1),
<a href="samtools-markdup.html"><em>samtools-markdup</em></a> (1),
<a href="samtools-merge.html"><em>samtools-merge</em></a> (1),
<a href="samtools-mpileup.html"><em>samtools-mpileup</em></a> (1),
<a href="samtools-phase.html"><em>samtools-phase</em></a> (1),
<a href="samtools-quickcheck.html"><em>samtools-quickcheck</em></a> (1),
<a href="samtools-reheader.html"><em>samtools-reheader</em></a> (1),
<a href="samtools-rmdup.html"><em>samtools-rmdup</em></a> (1),
<a href="samtools-sort.html"><em>samtools-sort</em></a> (1),
<a href="samtools-split.html"><em>samtools-split</em></a> (1),
<a href="samtools-stats.html"><em>samtools-stats</em></a> (1),
<a href="samtools-targetcut.html"><em>samtools-targetcut</em></a> (1),
<a href="samtools-tview.html"><em>samtools-tview</em></a> (1),
<a href="samtools-view.html"><em>samtools-view</em></a> (1),
<a href="bcftools.html"><em>bcftools</em></a> (1),
<a href="sam.html"><em>sam</em></a> (5),
<a href="tabix.html"><em>tabix</em></a> (1)
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
<br>
File format specification of SAM/BAM,CRAM,VCF/BCF: &lt;<a href="http://samtools.github.io/hts-specs">http://samtools.github.io/hts-specs</a>&gt;
<br>
Samtools latest source: &lt;<a href="https://github.com/samtools/samtools">https://github.com/samtools/samtools</a>&gt;
<br>
HTSlib latest source: &lt;<a href="https://github.com/samtools/htslib">https://github.com/samtools/htslib</a>&gt;
<br>
Bcftools website: &lt;<a href="http://samtools.github.io/bcftools">http://samtools.github.io/bcftools</a>&gt;
