---
permalink: /doc/samtools-bedcov.1.html
layout: manpage
title: samtools-bedcov(1) manual page
package: samtools-1.9
date: 14 August 2018
section_text: Bioinformatics tools
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools bedcov &ndash; reports coverage over regions in a supplied BED file
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
samtools <a href="samtools-bedcov.1.html">bedcov</a>
[<em>options</em>]
<em>region.bed</em> <em>in1.sam</em>|<em>in1.bam</em>|<em>in1.cram</em>[...]
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
<p>
Reports the total read base count (i.e. the sum of per base read depths)
for each genomic region specified in the supplied BED file. The regions
are output as they appear in the BED file and are 0-based.
Counts for each alignment file supplied are reported in separate columns.
<p>
<h1 id="OPTIONS"><a href="#OPTIONS">OPTIONS</a></h1>
<dl><dt><b>-Q </b><em>INT</em></dt><dd><p>
Only count reads with mapping quality greater than <em>INT</em>
</dd><dt><b>-j</b></dt><dd><p>
Do not include deletions (D) and ref skips (N) in bedcov computation.
</dd><dt><b>-X</b></dt><dd><p>
If this option is set, it will allows user to specify customized index file location(s) if the data
folder does not contain any index file. Example usage: samtools bedcov [options] -X &lt;in.bed&gt; &lt;/data_folder/in1.bam&gt; [...] &lt;/index_folder/index1.bai&gt; [...]
<p>
</dd></dl>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Written by Heng Li from the Sanger Institute.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<a href="samtools.1.html"><em>samtools</em></a> (1)
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
