---
permalink: /doc/samtools.html
layout: manpage
title: samtools(1) manual page
package: samtools-1.9
date: 18 July 2018
section_text: Bioinformatics tools
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools &ndash; Utilities for the Sequence Alignment/Map (SAM) format
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
samtools view -bt ref_list.txt -o aln.bam aln.sam.gz
<p>
samtools sort -T /tmp/aln.sorted -o aln.sorted.bam aln.bam
<p>
samtools index aln.sorted.bam
<p>
samtools idxstats aln.sorted.bam
<p>
samtools flagstat aln.sorted.bam
<p>
samtools stats aln.sorted.bam
<p>
samtools bedcov aln.sorted.bam
<p>
samtools depth aln.sorted.bam
<p>
samtools view aln.sorted.bam chr2:20,100,000-20,200,000
<p>
samtools merge out.bam in1.bam in2.bam in3.bam
<p>
samtools faidx ref.fasta
<p>
samtools fqidx ref.fastq
<p>
samtools tview aln.sorted.bam ref.fasta
<p>
samtools split merged.bam
<p>
samtools quickcheck in1.bam in2.cram
<p>
samtools dict -a GRCh38 -s "Homo sapiens" ref.fasta
<p>
samtools fixmate in.namesorted.sam out.bam
<p>
samtools mpileup -C50 -f ref.fasta -r chr3:1,000-2,000 in1.bam in2.bam
<p>
samtools flags PAIRED,UNMAP,MUNMAP
<p>
samtools fastq input.bam &gt; output.fastq
<p>
samtools fasta input.bam &gt; output.fasta
<p>
samtools addreplacerg -r 'ID:fish' -r 'LB:1334' -r 'SM:alpha' -o output.bam input.bam
<p>
samtools collate -o aln.name_collated.bam aln.sorted.bam
<p>
samtools depad input.bam
<p>
samtools markdup in.algnsorted.bam out.bam
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
Samtools is a set of utilities that manipulate alignments in the BAM
format. It imports from and exports to the SAM (Sequence Alignment/Map)
format, does sorting, merging and indexing, and allows to retrieve reads
in any regions swiftly.
<p>
Samtools is designed to work on a stream. It regards an input file `-'
as the standard input (stdin) and an output file `-' as the standard
output (stdout). Several commands can thus be combined with Unix
pipes. Samtools always output warning and error messages to the standard
error output (stderr).
<p>
Samtools is also able to open a BAM (not SAM) file on a remote FTP or
HTTP server if the BAM file name starts with `ftp://' or `http://'.
Samtools checks the current working directory for the index file and
will download the index upon absence. Samtools does not retrieve the
entire alignment file unless it is asked to do so.
<p>
<h1 id="COMMANDS_AND_OPTIONS"><a href="#COMMANDS_AND_OPTIONS">COMMANDS AND OPTIONS</a></h1>
<p>
<dl><dt><b>view</b></dt><dd><p>
samtools view
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[<em>region</em>...]
<p>
With no options or regions specified, prints all alignments in the specified
input alignment file (in SAM, BAM, or CRAM format) to standard output
in SAM format (with no header).
<p>
You may specify one or more space-separated region specifications after the
input filename to restrict output to only those alignments which overlap the
specified region(s). Use of region specifications requires a coordinate-sorted
and indexed input file (in BAM or CRAM format).
<p>
The
<b>-b</b>,
<b>-C</b>,
<b>-1</b>,
<b>-u</b>,
<b>-h</b>,
<b>-H</b>,
and
<b>-c</b>
options change the output format from the default of headerless SAM, and the
<b>-o</b>
and
<b>-U</b>
options set the output file name(s).
<p>
The
<b>-t</b>
and
<b>-T</b>
options provide additional reference data. One of these two options is required
when SAM input does not contain @SQ headers, and the
<b>-T</b>
option is required whenever writing CRAM output.
<p>
The
<b>-L</b>,
<b>-M</b>,
<b>-r</b>,
<b>-R</b>,
<b>-s</b>,
<b>-q</b>,
<b>-l</b>,
<b>-m</b>,
<b>-f</b>,
<b>-F</b>,
and
<b>-G</b>
options filter the alignments that will be included in the output to only those
alignments that match certain criteria.
<p>
The
<b>-x</b>
and
<b>-B</b>
options modify the data which is contained in each alignment.
<p>
Finally, the
<b>-@</b>
option can be used to allocate additional threads to be used for compression, and the
<b>-?</b>
option requests a long help message.
<p>
</dd><dt><b>REGIONS:</b></dt><dd><p>
Regions can be specified as: RNAME[:STARTPOS[-ENDPOS]] and all position
coordinates are 1-based.
<p>
Important note: when multiple regions are given, some alignments may be output
multiple times if they overlap more than one of the specified regions.
<p>
Examples of region specifications:
<dl><dt><b>chr1</b></dt><dd><p>
Output all alignments mapped to the reference sequence named `chr1' (i.e. @SQ SN:chr1).
</dd><dt><b>chr2:1000000</b></dt><dd><p>
The region on chr2 beginning at base position 1,000,000 and ending at the
end of the chromosome.
</dd><dt><b>chr3:1000-2000</b></dt><dd><p>
The 1001bp region on chr3 beginning at base position 1,000 and ending at base
position 2,000 (including both end positions).
</dd><dt><b>'*'</b></dt><dd><p>
Output the unmapped reads at the end of the file.
(This does not include any unmapped reads placed on a reference sequence
alongside their mapped mates.)
</dd><dt><b>.</b></dt><dd><p>
Output all alignments.
(Mostly unnecessary as not specifying a region at all has the same effect.)
</dd></dl>
<p>
<b>OPTIONS:</b>
<dl><dt><b>-b</b></dt><dd><p>
Output in the BAM format.
</dd><dt><b>-C</b></dt><dd><p>
Output in the CRAM format (requires -T).
</dd><dt><b>-1</b></dt><dd><p>
Enable fast BAM compression (implies -b).
</dd><dt><b>-u</b></dt><dd><p>
Output uncompressed BAM. This option saves time spent on
compression/decompression and is thus preferred when the output is piped
to another samtools command.
</dd><dt><b>-h</b></dt><dd><p>
Include the header in the output.
</dd><dt><b>-H</b></dt><dd><p>
Output the header only.
</dd><dt><b>-c</b></dt><dd><p>
Instead of printing the alignments, only count them and print the
total number. All filter options, such as
<b>-f</b>,
<b>-F</b>,
and
<b>-q</b>,
are taken into account.
</dd><dt><b>-?</b></dt><dd><p>
Output long help and exit immediately.
</dd><dt><b>-o </b><em>FILE</em></dt><dd><p>
Output to
<em>FILE [stdout].</em>
</dd><dt><b>-U </b><em>FILE</em></dt><dd><p>
Write alignments that are
<em>not</em>
selected by the various filter options to
<em>FILE</em>.
When this option is used, all alignments (or all alignments intersecting the
<em>regions</em>
specified) are written to either the output file or this file, but never both.
</dd><dt><b>-t </b><em>FILE</em></dt><dd><p>
A tab-delimited
<em>FILE</em>.
Each line must contain the reference name in the first column and the length of
the reference in the second column, with one line for each distinct reference.
Any additional fields beyond the second column are ignored. This file also
defines the order of the reference sequences in sorting. If you run:
`samtools faidx &lt;ref.fa&gt;', the resulting index file
<em>&lt;ref.fa&gt;.fai</em>
can be used as this
<em>FILE</em>.
</dd><dt><b>-T </b><em>FILE</em></dt><dd><p>
A FASTA format reference
<em>FILE</em>,
optionally compressed by
<b>bgzip</b>
and ideally indexed by
<b>samtools</b>
<b>faidx</b>.
If an index is not present, one will be generated for you.
</dd><dt><b>-L </b><em>FILE</em></dt><dd><p>
Only output alignments overlapping the input BED
<em>FILE</em>
[null].
</dd><dt><b>-M </b></dt><dd><p>
Use the multi-region iterator on the union of the BED file and
command-line region arguments.  This avoids re-reading the same regions
of files so can sometimes be much faster.  Note this also removes
duplicate sequences.  Without this a sequence that overlaps multiple
regions specified on the command line will be reported multiple times.
</dd><dt><b>-r </b><em>STR</em></dt><dd><p>
Output alignments in read group
<em>STR</em>
[null].
Note that records with no
<b>RG</b>
tag will also be output when using this option.
This behaviour may change in a future release.
</dd><dt><b>-R </b><em>FILE</em></dt><dd><p>
Output alignments in read groups listed in
<em>FILE</em>
[null].
Note that records with no
<b>RG</b>
tag will also be output when using this option.
This behaviour may change in a future release.
</dd><dt><b>-q </b><em>INT</em></dt><dd><p>
Skip alignments with MAPQ smaller than
<em>INT</em>
[0].
</dd><dt><b>-l </b><em>STR</em></dt><dd><p>
Only output alignments in library
<em>STR</em>
[null].
</dd><dt><b>-m </b><em>INT</em></dt><dd><p>
Only output alignments with number of CIGAR bases consuming query
sequence &ge;
<em>INT</em>
[0]
</dd><dt><b>-f </b><em>INT</em></dt><dd><p>
Only output alignments with all bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-F </b><em>INT</em></dt><dd><p>
Do not output alignments with any bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-G </b><em>INT</em></dt><dd><p>
Do not output alignments with all bits set in
<em>INT</em>
present in the FLAG field.  This is the opposite of <em>-f</em> such
that <em>-f12 -G12</em> is the same as no filtering at all.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-x </b><em>STR</em></dt><dd><p>
Read tag to exclude from output (repeatable) [null]
</dd><dt><b>-B</b></dt><dd><p>
Collapse the backward CIGAR operation.
</dd><dt><b>-s </b><em>FLOAT</em></dt><dd><p>
Output only a proportion of the input alignments.
This subsampling acts in the same way on all of the alignment records in
the same template or read pair, so it never keeps a read but not its mate.
<p>
The integer and fractional parts of the
<b>-s </b><em>INT</em><b>.</b><em>FRAC</em>
option are used separately: the part after the
decimal point sets the fraction of templates/pairs to be kept,
while the integer part is used as a seed that influences
<em>which</em>
subset of reads is kept.
<p>
When subsampling data that has previously been subsampled, be sure to use
a different seed value from those used previously; otherwise more reads
will be retained than expected.
</dd><dt><b>-@ </b><em>INT</em></dt><dd><p>
Number of BAM compression threads to use in addition to main thread [0].
</dd><dt><b>-S</b></dt><dd><p>
Ignored for compatibility with previous samtools versions.
Previously this option was required if input was in SAM format, but now the
correct format is automatically detected by examining the first few characters
of input.
</dd></dl>
<p>
</dd><dt><b>sort</b></dt><dd><p>
samtools sort
[<b>-l</b>
<em>level</em>]
[<b>-m</b>
<em>maxMem</em>]
[<b>-o</b>
<em>out.bam</em>]
[<b>-O</b>
<em>format</em>]
[<b>-n</b>]
[<b>-t</b>
<em>tag</em>]
[<b>-T</b>
<em>tmpprefix</em>]
[<b>-@</b>
<em>threads</em>] [<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>]
<p>
Sort alignments by leftmost coordinates, or by read name when
<b>-n</b>
is used.
An appropriate
<b>@HD-SO</b>
sort order header tag will be added or an existing one updated if necessary.
<p>
The sorted output is written to standard output by default, or to the
specified file
(<em>out.bam</em>)
when
<b>-o</b>
is used.
This command will also create temporary files
<em>tmpprefix</em><b>.</b><em>%d</em><b>.bam</b>
as needed when the entire alignment data cannot fit into memory
(as controlled via the
<b>-m</b>
option).
<p>
<b>Options:</b>
<dl><dt><b>-l </b><em>INT</em></dt><dd><p>
Set the desired compression level for the final output file, ranging from 0
(uncompressed) or 1 (fastest but minimal compression) to 9 (best compression
but slowest to write), similarly to
<b>gzip</b>(1)'s
compression level setting.
<p>
If
<b>-l</b>
is not used, the default compression level will apply.
</dd><dt><b>-m </b><em>INT</em></dt><dd><p>
Approximately the maximum required memory per thread, specified either in bytes
or with a
<b>K</b>, <b>M</b>, or <b>G</b>
suffix.
[768 MiB]
<p>
To prevent sort from creating a huge number of temporary files, it enforces a
minimum value of 1M for this setting.
</dd><dt><b>-n</b></dt><dd><p>
Sort by read names (i.e., the
<b>QNAME</b>
field) rather than by chromosomal coordinates.
</dd><dt><b>-t </b><em>TAG</em></dt><dd><p>
Sort first by the value in the alignment tag TAG, then by position or name (if
also using <b>-n</b>).
<b>-o </b><em>FILE</em>
Write the final sorted output to
<em>FILE</em>,
rather than to standard output.
</dd><dt><b>-O </b><em>FORMAT</em></dt><dd><p>
Write the final output as
<b>sam</b>, <b>bam</b>, or <b>cram</b>.
<p>
By default, samtools tries to select a format based on the
<b>-o</b>
filename extension; if output is to standard output or no format can be
deduced,
<b>bam</b>
is selected.
</dd><dt><b>-T </b><em>PREFIX</em></dt><dd><p>
Write temporary files to
<em>PREFIX</em><b>.</b><em>nnnn</em><b>.bam,</b>
or if the specified
<em>PREFIX</em>
is an existing directory, to
<em>PREFIX</em><b>/samtools.</b><em>mmm</em><b>.</b><em>mmm</em><b>.tmp.</b><em>nnnn</em><b>.bam,</b>
where
<em>mmm</em>
is unique to this invocation of the
<b>sort</b>
command.
<p>
By default, any temporary files are written alongside the output file, as
<em>out.bam</em><b>.tmp.</b><em>nnnn</em><b>.bam,</b>
or if output is to standard output, in the current directory as
<b>samtools.</b><em>mmm</em><b>.</b><em>mmm</em><b>.tmp.</b><em>nnnn</em><b>.bam.</b>
</dd><dt><b>-@ </b><em>INT</em></dt><dd><p>
Set number of sorting and compression threads.
By default, operation is single-threaded.
</dd></dl><p>
<b>Ordering Rules</b>
<p>
The following rules are used for ordering records.
<p>
If option <b>-t</b> is in use, records are first sorted by the value of
the given alignment tag, and then by position or name (if using <b>-n</b>).
For example, &ldquo;-t RG&rdquo; will make read group the primary sort key.  The
rules for ordering by tag are:
<p>
<ul><li>Records that do not have the tag are sorted before ones that do.
</li>
<li>If the types of the tags are different, they will be sorted so
that single character tags (type A) come before array tags (type B), then
string tags (types H and Z), then numeric tags (types f and i).
</li>
<li>Numeric tags (types f and i) are compared by value.  Note that comparisons
of floating-point values are subject to issues of rounding and precision.
</li>
<li>String tags (types H and Z) are compared based on the binary
contents of the tag using the C
<b>strcmp</b>(3)
function.
</li>
<li>Character tags (type A) are compared by binary character value.
</li>
<li>No attempt is made to compare tags of other types &mdash; notably type B
array values will not be compared.
</ul><p>
When the <b>-n</b> option is present, records are sorted by name.  Names are
compared so as to give a &ldquo;natural&rdquo; ordering &mdash; i.e. sections
consisting of digits are compared numerically while all other sections are
compared based on their binary representation.  This means &ldquo;a1&rdquo; will
come before &ldquo;b1&rdquo; and &ldquo;a9&rdquo; will come before &ldquo;a10&rdquo;.
Records with the same name will be ordered according to the values of
the READ1 and READ2 flags (see
<b>flags</b>).
<p>
When the <b>-n</b> option is
<b>not</b>
present, reads are sorted by reference (according to the order of the @SQ
header records), then by position in the reference, and then by the REVERSE
flag.
<p>
<b>Note</b>
<p>
<p>
Historically
<b>samtools sort</b>
also accepted a less flexible way of specifying the final and
temporary output filenames:
<p>
samtools sort
[<b>-f</b>] [<b>-o</b>]
<em>in.bam out.prefix</em>
<p>
This has now been removed.
The previous <em>out.prefix</em> argument (and <b>-f</b> option, if any)
should be changed to an appropriate combination of <b>-T</b> <em>PREFIX</em>
and <b>-o</b> <em>FILE</em>.  The previous <b>-o</b> option should be removed,
as output defaults to standard output.
<p>
</dd><dt><b>index</b></dt><dd><p>
samtools index
[<b>-bc</b>]
[<b>-m</b>
<em>INT</em>]
<em>aln.bam</em>|<em>aln.cram</em>
[<em>out.index</em>]
<p>
Index a coordinate-sorted BAM or CRAM file for fast random access.
(Note that this does not work with SAM files even if they are bgzip
compressed &mdash; to index such files, use tabix(1) instead.)
<p>
This index is needed when
<em>region</em>
arguments are used to limit
<b>samtools view</b>
and similar commands to particular regions of interest.
<p>
If an output filename is given, the index file will be written to
<em>out.index</em>.
Otherwise, for a CRAM file
<em>aln.cram</em>,
index file
<em>aln.cram</em><b>.crai</b>
will be created; for a BAM file
<em>aln.bam</em>,
either
<em>aln.bam</em><b>.bai</b>
or
<em>aln.bam</em><b>.csi</b>
will be created, depending on the index format selected.
<p>
<b>Options:</b>
<dl><dt><b>-b</b></dt><dd><p>
Create a BAI index.
This is currently the default when no format options are used.
</dd><dt><b>-c</b></dt><dd><p>
Create a CSI index.
By default, the minimum interval size for the index is 2^14, which is the same
as the fixed value used by the BAI format.
</dd><dt><b>-m </b><em>INT</em></dt><dd><p>
Create a CSI index, with a minimum interval size of 2^INT.
</dd></dl>
<p>
</dd><dt><b>idxstats</b></dt><dd><p>
samtools idxstats
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
<p>
Retrieve and print stats in the index file corresponding to the input file.
Before calling idxstats, the input BAM file should be indexed by samtools index.
<p>
If run on a SAM or CRAM file or an unindexed BAM file, this command
will still produce the same summary statistics, but does so by reading
through the entire file.  This is far slower than using the BAM
indices.
<p>
The output is TAB-delimited with each line consisting of reference sequence
name, sequence length, # mapped reads and # unmapped reads. It is written to
stdout.
<p>
</dd><dt><b>flagstat</b></dt><dd><p>
samtools flagstat
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
<p>
Does a full pass through the input file to calculate and print statistics
to stdout.
<p>
Provides counts for each of 13 categories based primarily on bit flags in
the FLAG field. Each category in the output is broken down into QC pass and
QC fail, which is presented as "#PASS + #FAIL" followed by a description of
the category.
<p>
The first row of output gives the total number of reads that are QC pass and
fail (according to flag bit 0x200). For example:
<p>
  122 + 28 in total (QC-passed reads + QC-failed reads)
<p>
Which would indicate that there are a total of 150 reads in the input file,
122 of which are marked as QC pass and 28 of which are marked as "not passing
quality controls"
<p>
Following this, additional categories are given for reads which are:
<p>
<dl><dt>secondary</dt><dd><p>
0x100 bit set
</dd><dt>supplementary</dt><dd><p>
0x800 bit set
</dd><dt>duplicates</dt><dd><p>
0x400 bit set
</dd><dt>mapped</dt><dd><p>
0x4 bit not set
</dd><dt>paired in sequencing</dt><dd><p>
0x1 bit set
</dd><dt>read1</dt><dd><p>
both 0x1 and 0x40 bits set
</dd><dt>read2</dt><dd><p>
both 0x1 and 0x80 bits set
</dd><dt>properly paired</dt><dd><p>
both 0x1 and 0x2 bits set and 0x4 bit not set
</dd><dt>with itself and mate mapped</dt><dd><p>
0x1 bit set and neither 0x4 nor 0x8 bits set
</dd><dt>singletons</dt><dd><p>
both 0x1 and 0x8 bits set and bit 0x4 not set
</dd></dl>
<p>
And finally, two rows are given that additionally filter on the reference
name (RNAME), mate reference name (MRNM), and mapping quality (MAPQ) fields:
<p>
<dl><dt>with mate mapped to a different chr</dt><dd><p>
0x1 bit set and neither 0x4 nor 0x8 bits set and MRNM not equal to RNAME
</dd><dt>with mate mapped to a different chr (mapQ&gt;=5)</dt><dd><p>
0x1 bit set and neither 0x4 nor 0x8 bits set
and MRNM not equal to RNAME and MAPQ &gt;= 5
</dd></dl>
<p>
</dd><dt><b>stats</b></dt><dd><p>
samtools stats
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[<em>region</em>...]
<p>
samtools stats collects statistics from BAM files and outputs in a text format.
The output can be visualized graphically using plot-bamstats.
<p>
<b>Options:</b>
<dl><dt><b>-c, --coverage </b><em>MIN</em><b>,</b><em>MAX</em><b>,</b><em>STEP</em></dt><dd><p>
Set coverage distribution to the specified range (MIN, MAX, STEP all given as integers)
[1,1000,1]
</dd><dt><b>-d, --remove-dups</b></dt><dd><p>
Exclude from statistics reads marked as duplicates
</dd><dt><b>-f, --required-flag </b><em>STR</em><b>|</b><em>INT</em></dt><dd><p>
Required flag, 0 for unset. See also `samtools flags`
[0]
</dd><dt><b>-F, --filtering-flag </b><em>STR</em><b>|</b><em>INT</em></dt><dd><p>
Filtering flag, 0 for unset. See also `samtools flags`
[0]
</dd><dt><b>--GC-depth </b><em>FLOAT</em></dt><dd><p>
the size of GC-depth bins (decreasing bin size increases memory requirement)
[2e4]
</dd><dt><b>-h, --help</b></dt><dd><p>
This help message
</dd><dt><b>-i, --insert-size </b><em>INT</em></dt><dd><p>
Maximum insert size
[8000]
</dd><dt><b>-I, --id </b><em>STR</em></dt><dd><p>
Include only listed read group or sample name
[]
</dd><dt><b>-l, --read-length </b><em>INT</em></dt><dd><p>
Include in the statistics only reads with the given read length
[-1]
</dd><dt><b>-m, --most-inserts </b><em>FLOAT</em></dt><dd><p>
Report only the main part of inserts
[0.99]
</dd><dt><b>-P, --split-prefix </b><em>STR</em></dt><dd><p>
A path or string prefix to prepend to filenames output when creating
categorised statistics files with
<b>-S</b>/<b>--split</b>.
[input filename]
</dd><dt><b>-q, --trim-quality </b><em>INT</em></dt><dd><p>
The BWA trimming parameter
[0]
</dd><dt><b>-r, --ref-seq </b><em>FILE</em></dt><dd><p>
Reference sequence (required for GC-depth and mismatches-per-cycle calculation).
[]
</dd><dt><b>-S, --split </b><em>TAG</em></dt><dd><p>
In addition to the complete statistics, also output categorised statistics
based on the tagged field
<em>TAG</em>
(e.g., use
<b>--split RG</b>
to split into read groups).
<p>
Categorised statistics are written to files named
&lt;<em>prefix</em>&gt;_&lt;<em>value</em>&gt;.bamstat,
where
<em>prefix</em>
is as given by
<b>--split-prefix</b>
(or the input filename by default) and
<em>value</em>
has been encountered as the specified tagged field's value in one or more
alignment records.
</dd><dt><b>-t, --target-regions </b><em>FILE</em></dt><dd><p>
Do stats in these regions only. Tab-delimited file chr,from,to, 1-based, inclusive.
[]
</dd><dt><b>-x, --sparse</b></dt><dd><p>
Suppress outputting IS rows where there are no insertions.
</dd><dt><b>-p, --remove-overlaps </b></dt><dd><p>
Remove overlaps of paired-end reads from coverage and base count computations.
</dd><dt><b>-g, --cov-threshold </b><em>INT</em></dt><dd><p>
Only bases with coverage above this value will be included in the target percentage computation [0]
</dd></dl>
<p>
</dd><dt><b>bedcov</b></dt><dd><p>
samtools bedcov
[<em>options</em>]
<em>region.bed</em> <em>in1.sam</em>|<em>in1.bam</em>|<em>in1.cram</em>[...]
<p>
Reports the total read base count (i.e. the sum of per base read depths)
for each genomic region specified in the supplied BED file. The regions
are output as they appear in the BED file and are 0-based.
Counts for each alignment file supplied are reported in separate columns.
<p>
<b>Options:</b>
<dl><dt><b>-Q </b><em>INT</em></dt><dd><p>
Only count reads with mapping quality greater than <em>INT</em>
</dd><dt><b>-j</b></dt><dd><p>
Do not include deletions (D) and ref skips (N) in bedcov computation.
</dd></dl>
<p>
</dd><dt><b>depth</b></dt><dd><p>
samtools depth
[<em>options</em>]
[<em>in1.sam</em>|<em>in1.bam</em>|<em>in1.cram</em> [<em>in2.sam</em>|<em>in2.bam</em>|<em>in2.cram</em>] [...]]
<p>
Computes the depth at each position or region.
<p>
<b>Options:</b>
<dl><dt><b>-a</b></dt><dd><p>
Output all positions (including those with zero depth)
</dd><dt><b>-a -a, -aa</b></dt><dd><p>
Output absolutely all positions, including unused reference sequences.
Note that when used in conjunction with a BED file the -a option may
sometimes operate as if -aa was specified if the reference sequence
has coverage outside of the region specified in the BED file.
</dd><dt><b>-b </b><em>FILE</em></dt><dd><p>
Compute depth at list of positions or regions in specified BED <em>FILE.</em>
[]
</dd><dt><b>-f </b><em>FILE</em></dt><dd><p>
Use the BAM files specified in the <em>FILE</em>
(a file of filenames, one file per line)
[]
</dd><dt><b>-l </b><em>INT</em></dt><dd><p>
Ignore reads shorter than <em>INT</em>
</dd><dt><b>-m, -d </b><em>INT</em></dt><dd><p>
Truncate reported depth at a maximum of <em>INT</em> reads.
[8000]. If 0, depth is set to the maximum integer value, effectively removing any depth limit.
</dd><dt><b>-q </b><em>INT</em></dt><dd><p>
Only count reads with base quality greater than <em>INT</em>
</dd><dt><b>-Q </b><em>INT</em></dt><dd><p>
Only count reads with mapping quality greater than <em>INT</em>
</dd><dt><b>-r </b><em>CHR</em><b>:</b><em>FROM</em><b>-</b><em>TO</em></dt><dd><p>
Only report depth in specified region.
</dd></dl>
<p>
</dd><dt><b>merge</b></dt><dd><p>
samtools merge [-nur1f] [-h inh.sam] [-R reg] [-b &lt;list&gt;] &lt;out.bam&gt; &lt;in1.bam&gt; [&lt;in2.bam&gt; &lt;in3.bam&gt; ... &lt;inN.bam&gt;]
<p>
Merge multiple sorted alignment files, producing a single sorted output file
that contains all the input records and maintains the existing sort order.
<p>
If
<b>-h</b>
is specified the @SQ headers of input files will be merged into the specified header, otherwise they will be merged
into a composite header created from the input headers.  If in the process of merging @SQ lines for coordinate sorted
input files, a conflict arises as to the order (for example input1.bam has @SQ for a,b,c and input2.bam has b,a,c)
then the resulting output file will need to be re-sorted back into coordinate order.
<p>
Unless the
<b>-c</b>
or
<b>-p</b>
flags are specified then when merging @RG and @PG records into the output header then any IDs found to be duplicates
of existing IDs in the output header will have a suffix appended to them to differentiate them from similar header
records from other files and the read records will be updated to reflect this.
<p>
The ordering of the records in the input files must match the usage of the
<b>-n</b> and <b>-t</b> command-line options.  If they do not, the output
order will be undefined.  See
<b>sort</b>
for information about record ordering.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-1</b></dt><dd><p>
Use zlib compression level 1 to compress the output.
</dd><dt><b>-b</b><em> FILE</em></dt><dd><p>
List of input BAM files, one file per line.
</dd><dt><b>-f</b></dt><dd><p>
Force to overwrite the output file if present.
</dd><dt><b>-h</b><em> FILE</em></dt><dd><p>
Use the lines of
<em>FILE</em>
as `@' headers to be copied to
<em>out.bam</em>,
replacing any header lines that would otherwise be copied from
<em>in1.bam</em>.
(<em>FILE</em>
is actually in SAM format, though any alignment records it may contain
are ignored.)
</dd><dt><b>-n</b></dt><dd><p>
The input alignments are sorted by read names rather than by chromosomal
coordinates
</dd><dt><b>-t TAG</b></dt><dd><p>
The input alignments have been sorted by the value of TAG, then by either
position or name (if <b>-n</b> is given).
</dd><dt><b>-R</b><em> STR</em></dt><dd><p>
Merge files in the specified region indicated by
<em>STR</em>
[null]
</dd><dt><b>-r</b></dt><dd><p>
Attach an RG tag to each alignment. The tag value is inferred from file names.
</dd><dt><b>-u</b></dt><dd><p>
Uncompressed BAM output
</dd><dt><b>-c</b></dt><dd><p>
When several input files contain @RG headers with the same ID, emit only one
of them (namely, the header line from the first file we find that ID in) to
the merged output file.
Combining these similar headers is usually the right thing to do when the
files being merged originated from the same file.
<p>
Without <b>-c</b>, all @RG headers appear in the output file, with random
suffixes added to their IDs where necessary to differentiate them.
</dd><dt><b>-p</b></dt><dd><p>
Similarly, for each @PG ID in the set of files to merge, use the @PG line
of the first file we find that ID in rather than adding a suffix to
differentiate similar IDs.
</dd></dl>
<p>
</dd><dt><b>faidx</b></dt><dd><p>
samtools faidx &lt;ref.fasta&gt; [region1 [...]]
<p>
Index reference sequence in the FASTA format or extract subsequence from
indexed reference sequence. If no region is specified,
<b>faidx</b>
will index the file and create
<em>&lt;ref.fasta&gt;.fai</em>
on the disk. If regions are specified, the subsequences will be
retrieved and printed to stdout in the FASTA format.
<p>
The input file can be compressed in the
<b>BGZF</b>
format.
<p>
The sequences in the input file should all have different names.
If they do not, indexing will emit a warning about duplicate sequences and
retrieval will only produce subsequences from the first sequence with the
duplicated name.
<p>
FASTQ files can be read and indexed by this command.  Without using
<b>--fastq</b>
any extracted subsequence will be in FASTA format.
<p>
<b>Options</b>
<dl><dt><b>-o, --output </b><em>FILE</em></dt><dd><p>
Write FASTA to file rather than to stdout.
</dd><dt><b>-n, --length </b><em>INT</em></dt><dd><p>
Length of FASTA sequence line.
[60]
</dd><dt><b>-c, --continue</b></dt><dd><p>
Continue working if a non-existant region is requested.
</dd><dt><b>-r, --region-file </b><em>FILE</em></dt><dd><p>
Read regions from a file. Format is chr:from-to, one per line.
</dd><dt><b>-f, --fastq</b></dt><dd><p>
Read FASTQ files and output extracted sequences in FASTQ format.  Same as using samtools fqidx.
</dd><dt><b>-i, --reverse-complement</b></dt><dd><p>
Output the sequence as the reverse complement.
When this option is used, &ldquo;/rc&rdquo; will be appended to the sequence names.
To turn this off or change the string appended, use the
<b>--mark-strand</b>
option.
</dd><dt><b>--mark-strand TYPE</b></dt><dd><p>
Append strand indicator to sequence name.  TYPE can be one of:
<dl><dt><b>rc</b></dt><dd><p>
Append '/rc' when writing the reverse complement.  This is the default.
</dd><dt><b>no</b></dt><dd><p>
Do not append anything.
</dd><dt><b>sign</b></dt><dd><p>
Append '(+)' for forward strand or '(-)' for reverse complement.  This matches
the output of &ldquo;bedtools getfasta -s&rdquo;.
</dd><dt><b>custom,&lt;pos&gt;,&lt;neg&gt;</b></dt><dd><p>
Append string &lt;pos&gt; to names when writing the forward strand and &lt;neg&gt; when
writing the reverse strand.  Spaces are preserved, so it is possible to move
the indicator into the comment part of the description line by including
a leading space in the strings &lt;pos&gt; and &lt;neg&gt;.
</dd></dl>
</dd><dt><b>-h, --help</b></dt><dd><p>
Print help message and exit.
</dd></dl>
<p>
</dd><dt><b>fqidx</b></dt><dd><p>
samtools fqidx &lt;ref.fastq&gt; [region1 [...]]
<p>
Index reference sequence in the FASTQ format or extract subsequence from
indexed reference sequence. If no region is specified,
<b>fqidx</b>
will index the file and create
<em>&lt;ref.fastq&gt;.fai</em>
on the disk. If regions are specified, the subsequences will be
retrieved and printed to stdout in the FASTQ format.
<p>
The input file can be compressed in the
<b>BGZF</b>
format.
<p>
The sequences in the input file should all have different names.
If they do not, indexing will emit a warning about duplicate sequences and
retrieval will only produce subsequences from the first sequence with the
duplicated name.
<p>
<b>samtools fqidx</b>
should only be used on fastq files with a small number of entries.
Trying to use it on a file containing millions of short sequencing reads
will produce an index that is almost as big as the original file, and
searches using the index will be very slow and use a lot of memory.
<p>
<b>Options</b>
<dl><dt><b>-o, --output </b><em>FILE</em></dt><dd><p>
Write FASTQ to file rather than to stdout.
</dd><dt><b>-n, --length </b><em>INT</em></dt><dd><p>
Length of FASTQ sequence line.
[60]
</dd><dt><b>-c, --continue</b></dt><dd><p>
Continue working if a non-existant region is requested.
</dd><dt><b>-r, --region-file </b><em>FILE</em></dt><dd><p>
Read regions from a file. Format is chr:from-to, one per line.
</dd><dt><b>-i, --reverse-complement</b></dt><dd><p>
Output the sequence as the reverse complement.
When this option is used, &ldquo;/rc&rdquo; will be appended to the sequence names.
To turn this off or change the string appended, use the
<b>--mark-strand</b>
option.
</dd><dt><b>--mark-strand TYPE</b></dt><dd><p>
Append strand indicator to sequence name.  TYPE can be one of:
<dl><dt><b>rc</b></dt><dd><p>
Append '/rc' when writing the reverse complement.  This is the default.
</dd><dt><b>no</b></dt><dd><p>
Do not append anything.
</dd><dt><b>sign</b></dt><dd><p>
Append '(+)' for forward strand or '(-)' for reverse complement.  This matches
the output of &ldquo;bedtools getfasta -s&rdquo;.
</dd><dt><b>custom,&lt;pos&gt;,&lt;neg&gt;</b></dt><dd><p>
Append string &lt;pos&gt; to names when writing the forward strand and &lt;neg&gt; when
writing the reverse strand.  Spaces are preserved, so it is possible to move
the indicator into the comment part of the description line by including
a leading space in the strings &lt;pos&gt; and &lt;neg&gt;.
</dd></dl>
</dd><dt><b>-h, --help</b></dt><dd><p>
Print help message and exit.
</dd></dl>
<p>
</dd><dt><b>tview</b></dt><dd><p>
samtools tview
[<b>-p</b>
<em>chr:pos</em>]
[<b>-s</b>
<em>STR</em>]
[<b>-d</b>
<em>display</em>]
&lt;in.sorted.bam&gt;
[ref.fasta]
<p>
Text alignment viewer (based on the ncurses library). In the viewer,
press `?' for help and press `g' to check the alignment start from a
region in the format like `chr10:10,000,000' or `=10,000,000' when
viewing the same reference sequence.
<p>
<b>Options:</b>
<dl><dt><b>-d</b><em> display</em></dt><dd><p>
Output as (H)tml or (C)urses or (T)ext
</dd><dt><b>-p</b><em> chr:pos</em></dt><dd><p>
Go directly to this position
</dd><dt><b>-s</b><em> STR</em></dt><dd><p>
Display only alignments from this sample or read group
</dd></dl>
<p>
</dd><dt><b>split</b></dt><dd><p>
samtools split
[<em>options</em>]
<em>merged.sam</em>|<em>merged.bam</em>|<em>merged.cram</em>
<p>
Splits a file by read group.
<p>
<b>Options:</b>
<dl><dt><b>-u </b><em>FILE1</em></dt><dd><p>
Put reads with no RG tag or an unrecognised RG tag into <em>FILE1</em>
</dd><dt><b>-u </b><em>FILE1</em><b>:</b><em>FILE2</em></dt><dd><p>
As above, but assigns an RG tag as given in the header of <em>FILE2</em>
</dd><dt><b>-f </b><em>STRING</em></dt><dd><p>
Output filename format string (see below)
["%*_%#.%."]
</dd><dt><b>-v</b></dt><dd><p>
Verbose output
</dd></dl><p>
Format string expansions:
<table align="center">
<tr><td>%%</td><td>%</td></tr>
<tr><td>%*</td><td>basename</td></tr>
<tr><td>%#</td><td>@RG index</td></tr>
<tr><td>%!</td><td>@RG ID</td></tr>
<tr><td>%.</td><td>output format filename extension</td></tr>
</table>
<p>
</dd><dt><b>quickcheck</b></dt><dd><p>
samtools quickcheck
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em>
[ ... ]
<p>
Quickly check that input files appear to be intact. Checks that beginning of the
file contains a valid header (all formats) containing at least one target
sequence and then seeks to the end of the file and checks that an end-of-file
(EOF) is present and intact (BAM only).
<p>
Data in the middle of the file is not read since that would be much more time
consuming, so please note that this command will not detect internal corruption,
but is useful for testing that files are not truncated before performing more
intensive tasks on them.
<p>
This command will exit with a non-zero exit code if any input files don't have a
valid header or are missing an EOF block. Otherwise it will exit successfully
(with a zero exit code).
<p>
<b>Options:</b>
<dl><dt><b>-v</b></dt><dd><p>
Verbose output: will additionally print the names of all input files that don't
pass the check to stdout. Multiple -v options will cause additional messages
regarding check results to be printed to stderr.
</dd><dt><b>-q</b></dt><dd><p>
Quiet mode: disables warning messages on stderr about files that fail.
If both -q and -v options are used then the appropriate level of -v takes precedence.
</dd></dl>
<p>
</dd><dt><b>dict</b></dt><dd><p>
samtools dict &lt;ref.fasta|ref.fasta.gz&gt;
<p>
Create a sequence dictionary file from a fasta file.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-a, --assembly</b><em> STR</em></dt><dd><p>
Specify the assembly for the AS tag.
</dd><dt><b>-H, --no-header</b></dt><dd><p>
Do not print the @HD header line.
</dd><dt><b>-o, --output</b><em> FILE</em></dt><dd><p>
Output to
<em>FILE</em>
[stdout].
</dd><dt><b>-s, --species</b><em> STR</em></dt><dd><p>
Specify the species for the SP tag.
</dd><dt><b>-u, --uri</b><em> STR</em></dt><dd><p>
Specify the URI for the UR tag. Defaults to
the absolute path of
<em>ref.fasta</em>
unless reading from stdin.
</dd></dl>
<p>
</dd><dt><b>fixmate</b></dt><dd><p>
samtools fixmate
[<b>-rpcm</b>]
[<b>-O</b>
<em>format</em>]
<em>in.nameSrt.bam out.bam</em>
<p>
Fill in mate coordinates, ISIZE and mate related flags from a
name-sorted alignment.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-r</b></dt><dd><p>
Remove secondary and unmapped reads.
</dd><dt><b>-p</b></dt><dd><p>
Disable FR proper pair check.
</dd><dt><b>-c</b></dt><dd><p>
Add template cigar ct tag.
</dd><dt><b>-m</b></dt><dd><p>
Add ms (mate score) tags.  These are used by
<b>markdup</b>
to select the best reads to keep.
</dd><dt><b>-O </b><em>FORMAT</em></dt><dd><p>
Write the final output as
<b>sam</b>, <b>bam</b>, or <b>cram</b>.
<p>
By default, samtools tries to select a format based on the output
filename extension; if output is to standard output or no format can be
deduced,
<b>bam</b>
is selected.
</dd></dl>
<p>
</dd><dt><b>mpileup</b></dt><dd><p>
samtools mpileup
[<b>-EB</b>]
[<b>-C</b>
<em>capQcoef</em>]
[<b>-r</b>
<em>reg</em>]
[<b>-f</b>
<em>in.fa</em>]
[<b>-l</b>
<em>list</em>]
[<b>-Q</b>
<em>minBaseQ</em>]
[<b>-q</b>
<em>minMapQ</em>]
<em>in.bam</em>
[<em>in2.bam</em>
[<em>...</em>]]
<p>
Generate pileup for one or multiple BAM files. Alignment records
are grouped by sample (SM) identifiers in @RG header lines. If sample
identifiers are absent, each input file is regarded as one sample.
<p>
Samtools mpileup can still produce VCF and BCF output, but this feature is
deprecated and will be removed in a future release.  Please use
<b>bcftools mpileup</b>
for this instead.  (Documentation on the deprecated options has been removed
from this manual page, but older versions are available online
at &lt;<a href="http://www.htslib.org/doc/">http://www.htslib.org/doc/</a>&gt;.)
<p>
In the pileup format (without
<b>-u</b> or<b> -g</b>),
each
line represents a genomic position, consisting of chromosome name,
1-based coordinate, reference base, the number of reads covering the site,
read bases, base qualities and alignment
mapping qualities. Information on match, mismatch, indel, strand,
mapping quality and start and end of a read are all encoded at the read
base column. At this column, a dot stands for a match to the reference
base on the forward strand, a comma for a match on the reverse strand,
a '&gt;' or '&lt;' for a reference skip, `ACGTN' for a mismatch on the forward
strand and `acgtn' for a mismatch on the reverse strand. A pattern
`\\+[0-9]+[ACGTNacgtn]+' indicates there is an insertion between this
reference position and the next reference position. The length of the
insertion is given by the integer in the pattern, followed by the
inserted sequence. Similarly, a pattern `-[0-9]+[ACGTNacgtn]+'
represents a deletion from the reference. The deleted bases will be
presented as `*' in the following lines. Also at the read base column, a
symbol `^' marks the start of a read. The ASCII of the character
following `^' minus 33 gives the mapping quality. A symbol `$' marks the
end of a read segment.
<p>
Note that there are two orthogonal ways to specify locations in the
input file; via <b>-r</b> <em>region</em> and <b>-l</b> <em>file</em>.  The
former uses (and requires) an index to do random access while the
latter streams through the file contents filtering out the specified
regions, requiring no index.  The two may be used in conjunction.  For
example a BED file containing locations of genes in chromosome 20
could be specified using <b>-r 20 -l chr20.bed</b>, meaning that the
index is used to find chromosome 20 and then it is filtered for the
regions listed in the bed file.
<p>
<b>Input Options:</b>
<dl><dt><b>-6, --illumina1.3+</b></dt><dd><p>
Assume the quality is in the Illumina 1.3+ encoding.
</dd><dt><b>-A, --count-orphans</b></dt><dd><p>
Do not skip anomalous read pairs in variant calling.
</dd><dt><b>-b, --bam-list</b><em> FILE</em></dt><dd><p>
List of input BAM files, one file per line [null]
</dd><dt><b>-B, --no-BAQ</b></dt><dd><p>
Disable base alignment quality (BAQ) computation.
See
<b>BAQ</b>
below.
</dd><dt><b>-C, --adjust-MQ</b><em> INT</em></dt><dd><p>
Coefficient for downgrading mapping quality for reads containing
excessive mismatches. Given a read with a phred-scaled probability q of
being generated from the mapped position, the new mapping quality is
about sqrt((INT-q)/INT)*INT. A zero value disables this
functionality; if enabled, the recommended value for BWA is 50. [0]
</dd><dt><b>-d, --max-depth</b><em> INT</em></dt><dd><p>
At a position, read maximally
<em>INT</em>
reads per input file. Setting this limit reduces the amount of memory and
time needed to process regions with very high coverage.  Passing zero for this
option sets it to the highest possible value, effectively removing the depth
limit. [8000]
<p>
Note that up to release 1.8, samtools would enforce a minimum value for
this option.  This no longer happens and the limit is set exactly as
specified.
</dd><dt><b>-E, --redo-BAQ</b></dt><dd><p>
Recalculate BAQ on the fly, ignore existing BQ tags.
See
<b>BAQ</b>
below.
</dd><dt><b>-f, --fasta-ref</b><em> FILE</em></dt><dd><p>
The
<b>faidx</b>-indexed
reference file in the FASTA format. The file can be optionally compressed by
<b>bgzip</b>.
[null]
<p>
Supplying a reference file will enable base alignment quality calculation
for all reads aligned to a reference in the file.  See
<b>BAQ</b>
below.
</dd><dt><b>-G, --exclude-RG</b><em> FILE</em></dt><dd><p>
Exclude reads from readgroups listed in FILE (one @RG-ID per line)
</dd><dt><b>-l, --positions</b><em> FILE</em></dt><dd><p>
BED or position list file containing a list of regions or sites where
pileup or BCF should be generated. Position list files contain two
columns (chromosome and position) and start counting from 1.  BED
files contain at least 3 columns (chromosome, start and end position)
and are 0-based half-open.
<br>
While it is possible to mix both position-list and BED coordinates in
the same file, this is strongly ill advised due to the differing
coordinate systems. [null]
</dd><dt><b>-q, -min-MQ</b><em> INT</em></dt><dd><p>
Minimum mapping quality for an alignment to be used [0]
</dd><dt><b>-Q, --min-BQ</b><em> INT</em></dt><dd><p>
Minimum base quality for a base to be considered [13]
</dd><dt><b>-r, --region</b><em> STR</em></dt><dd><p>
Only generate pileup in region. Requires the BAM files to be indexed.
If used in conjunction with -l then considers the intersection of the
two requests.
<em>STR</em>
[all sites]
</dd><dt><b>-R, --ignore-RG</b></dt><dd><p>
Ignore RG tags. Treat all reads in one BAM as one sample.
</dd><dt><b>--rf, --incl-flags</b><em> STR|INT</em></dt><dd><p>
Required flags: skip reads with mask bits unset [null]
</dd><dt><b>--ff, --excl-flags</b><em> STR|INT</em></dt><dd><p>
Filter flags: skip reads with mask bits set
[UNMAP,SECONDARY,QCFAIL,DUP]
</dd><dt><b>-x, --ignore-overlaps</b></dt><dd><p>
Disable read-pair overlap detection.
</dd></dl><p>
<b>Output Options:</b>
<dl><dt><b>-o, --output </b><em>FILE</em></dt><dd><p>
Write pileup output to
<em>FILE</em>,
rather than the default of standard output.
<p>
(The same short option is used for both the deprecated
<b>--open-prob</b>
option and
<b>--output .</b>
If
<b>-o</b>'s
argument contains any non-digit characters other than a leading + or - sign,
it is interpreted as
<b>--output</b>.
Usually the filename extension will take care of this, but to write to an
entirely numeric filename use
<b>-o ./123</b>
or
<b>--output 123</b>.)
</dd><dt><b>-O, --output-BP</b></dt><dd><p>
Output base positions on reads.
</dd><dt><b>-s, --output-MQ</b></dt><dd><p>
Output mapping quality.
</dd><dt><b>--output-QNAME</b></dt><dd><p>
Output an extra column containing comma-separated read names.
</dd><dt><b>-a</b></dt><dd><p>
Output all positions, including those with zero depth.
</dd><dt><b>-a -a, -aa</b></dt><dd><p>
Output absolutely all positions, including unused reference sequences.
Note that when used in conjunction with a BED file the -a option may
sometimes operate as if -aa was specified if the reference sequence
has coverage outside of the region specified in the BED file.
</dd></dl><p>
<b>BAQ (Base Alignment Quality)</b>
<p>
BAQ is the Phred-scaled probability of a read base being misaligned.
It greatly helps to reduce false SNPs caused by misalignments.
BAQ is calculated using the probabilistic realignment method described
in the paper &ldquo;Improving SNP discovery by base alignment quality&rdquo;,
Heng Li, Bioinformatics, Volume 27, Issue 8
&lt;<a href="https://doi.org/10">https://doi.org/10</a>.1093/bioinformatics/btr076&gt;
<p>
BAQ is turned on when a reference file is supplied using the
<b>-f</b>
option.  To disable it, use the
<b>-B</b>
option.
<p>
It is possible to store pre-calculated BAQ values in a SAM BQ:Z tag.
Samtools mpileup will use the precalculated values if it finds them.
The
<b>-E</b>
option can be used to make it ignore the contents of the BQ:Z tag and
force it to recalculate the BAQ scores by making a new alignment.
<p>
</dd><dt><b>flags</b></dt><dd><p>
samtools flags INT|STR[,...]
<p>
Convert between textual and numeric flag representation.
<p>
<b>FLAGS:</b>
<table>
<tr><td>0x1</td><td>PAIRED</td><td>paired-end (or multiple-segment) sequencing technology</td></tr>
<tr><td>0x2</td><td>PROPER_PAIR</td><td>each segment properly aligned according to the aligner</td></tr>
<tr><td>0x4</td><td>UNMAP</td><td>segment unmapped</td></tr>
<tr><td>0x8</td><td>MUNMAP</td><td>next segment in the template unmapped</td></tr>
<tr><td>0x10</td><td>REVERSE</td><td>SEQ is reverse complemented</td></tr>
<tr><td>0x20</td><td>MREVERSE</td><td>SEQ of the next segment in the template is reverse complemented</td></tr>
<tr><td>0x40</td><td>READ1</td><td>the first segment in the template</td></tr>
<tr><td>0x80</td><td>READ2</td><td>the last segment in the template</td></tr>
<tr><td>0x100</td><td>SECONDARY</td><td>secondary alignment</td></tr>
<tr><td>0x200</td><td>QCFAIL</td><td>not passing quality controls</td></tr>
<tr><td>0x400</td><td>DUP</td><td>PCR or optical duplicate</td></tr>
<tr><td>0x800</td><td>SUPPLEMENTARY</td><td>supplementary alignment</td></tr>
</table>
<p>
</dd><dt><b>fastq/a</b></dt><dd><p>
samtools fastq
[<em>options</em>]
<em>in.bam</em>
<br>
samtools fasta
[<em>options</em>]
<em>in.bam</em>
<p>
Converts a BAM or CRAM into either FASTQ or FASTA format depending on the
command invoked. The files will be automatically compressed if the
file names have a .gz or .bgzf extension.
<p>
The input to this program must be collated by name.
Use
<b>samtools collate</b>
or
<b>samtools sort -n</b>
to ensure this.
<p>
For each different QNAME, the input records are categorised according to
the state of the READ1 and READ2 flag bits.
The three categories used are:
<p>
1 : Only READ1 is set.
<p>
2 : Only READ2 is set.
<p>
0 : Either both READ1 and READ2 are set; or neither is set.
<p>
The exact meaning of these categories depends on the sequencing technology
used.
It is expected that ordinary single and paired-end sequencing reads will be
in categories 1 and 2 (in the case of paired-end reads, one read of the pair
will be in category 1, the other in category 2).
Category 0 is essentially a &ldquo;catch-all&rdquo; for reads that do not
fit into a simple paired-end sequencing model.
<p>
For each category only one sequence will be written for a given QNAME.
If more than one record is available for a given QNAME and category,
the first in input file order that has quality values will be used.
If none of the candidate records has quality values, then the first in
input file order will be used instead.
<p>
Sequences will be written to standard output unless one of the
<b>-1</b>,<b>-2</b>, or <b>-0</b>
options is used, in which case sequences for that category will be written to
the specified file.
<p>
If a singleton file is specified using the
<b>-s</b>
option then only paired sequences will be output for categories 1 and 2;
paired meaning that for a given QNAME there are sequences for both
category 1
<b>and</b>
2.
If there is a sequence for only one of categories 1 or 2 then it will be
diverted into the specified singletons file.
This can be used to prepare fastq files for programs that cannot handle
a mixture of paired and singleton reads.
<p>
The
<b>-s</b>
option only affects category 1 and 2 records.
The output for category 0 will be the same irrespective of the use of this
option.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-n</b></dt><dd><p>
By default, either '/1' or '/2' is added to the end of read names
where the corresponding READ1 or READ2 FLAG bit is set.
Using
<b>-n</b>
causes read names to be left as they are.
</dd><dt><b>-N</b></dt><dd><p>
Always add either '/1' or '/2' to the end of read names
even when put into different files.
</dd><dt><b>-O</b></dt><dd><p>
Use quality values from OQ tags in preference to standard quality string
if available.
</dd><dt><b>-s FILE</b></dt><dd><p>
Write singleton reads to FILE.
</dd><dt><b>-t</b></dt><dd><p>
Copy RG, BC and QT tags to the FASTQ header line, if they exist.
</dd><dt><b>-T TAGLIST</b></dt><dd><p>
Specify a comma-separated list of tags to copy to the FASTQ header line, if
they exist.
</dd><dt><b>-1 FILE</b></dt><dd><p>
Write reads with the READ1 FLAG set (and READ2 not set) to FILE instead of
outputting them.
If the
<b>-s</b>
option is used, only paired reads will be written to this file.
</dd><dt><b>-2 FILE</b></dt><dd><p>
Write reads with the READ2 FLAG set (and READ1 not set) to FILE instead of
outputting them.
If the
<b>-s</b>
option is used, only paired reads will be written to this file.
</dd><dt><b>-0 FILE</b></dt><dd><p>
Write reads where the READ1 and READ2 FLAG bits set are either both set
or both unset to FILE instead of outputting them.
</dd><dt><b>-f </b><em>INT</em></dt><dd><p>
Only output alignments with all bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-F </b><em>INT</em></dt><dd><p>
Do not output alignments with any bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-G </b><em>INT</em></dt><dd><p>
Only EXCLUDE reads with all of the bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-i</b></dt><dd><p>
add Illumina Casava 1.8 format entry to header (eg 1:N:0:ATCACG)
</dd><dt><b>-c [0..9]</b></dt><dd><p>
set compression level when writing gz or bgzf fastq files.
</dd><dt><b>--i1 FILE</b></dt><dd><p>
write first index reads to FILE
</dd><dt><b>--i2 FILE</b></dt><dd><p>
write second index reads to FILE
</dd><dt><b>--barcode-tag TAG</b></dt><dd><p>
aux tag to find index reads in [default: BC]
</dd><dt><b>--quality-tag TAG</b></dt><dd><p>
aux tag to find index quality in [default: QT]
</dd><dt><b>--index-format STR</b></dt><dd><p>
string to describe how to parse the barcode and quality tags. For example:
<p>
<dl><dt><b>i14i8</b></dt><dd><p>
the first 14 characters are index 1, the next 8 characters are index 2
</dd><dt><b>n8i14</b></dt><dd><p>
ignore the first 8 characters, and use the next 14 characters for index 1
<p>
If the tag contains a separator, then the numeric part can be replaced with '*' to
mean 'read until the separator or end of tag', for example:
</dd><dt><b>n*i*</b></dt><dd><p>
ignore the left part of the tag until the separator, then use the second part
</dd></dl>
<p>
<b>EXAMPLES</b>
<p>
Output paired reads to separate files, discarding singletons, supplementary
and secondary reads.
The resulting files can be used with, for example, the
<b>bwa</b>
aligner.
<pre>
samtools fastq -1 paired1.fq -2 paired2.fq -0 /dev/null -s /dev/null -n -F 0x900 in.bam
</pre>
<p>
Output paired and singleton reads in a single file, discarding supplementary
and secondary reads.
To get all of the reads in a single file, it is necessary to redirect the
output of samtools fastq.
The output file is suitable for use with
<b>bwa mem -p</b>
which understands interleaved files containing a mixture of paired and
singleton reads.
<pre>
samtools fastq -0 /dev/null -F 0x900 in.bam &gt; all_reads.fq
</pre>
<p>
Output paired reads in a single file, discarding supplementary and
secondary reads.
Save any singletons in a separate file.
Append /1 and /2 to read names.
This format is suitable for use by
<b>NextGenMap</b>
when using its
<b>-p</b> and <b>-q</b> options.
With this aligner, paired reads must be mapped separately to the singletons.
<pre>
samtools fastq -0 /dev/null -s single.fq -N -F 0x900 in.bam &gt; paired.fq
</pre>
<p>
<b>BUGS</b>
<p>
<ul><li>The way of specifying output files is far to complicated and easy to get wrong.
<p>
</li>
<li>The default value for the -F option should really be 0x900 so that secondary
and supplementary reads are automatically excluded.
The existing default of 0 is retained for reasons of compatibility.
<p>
<p>
</dd><dt><b>collate</b></dt><dd><p>
samtools collate
[<em>options</em>]
<em>in.sam</em>|<em>in.bam</em>|<em>in.cram</em> [<em>&lt;prefix&gt;</em>]
<p>
Shuffles and groups reads together by their names.
A faster alternative to a full query name sort,
<b>collate</b>
ensures that reads of the same name are grouped together in contiguous groups,
but doesn't make any guarantees about the order of read names between groups.
<p>
The output from this command should be suitable for any operation that
requires all reads from the same template to be grouped together.
<p>
If present, &lt;prefix&gt; is used to name the temporary files that collate
uses when sorting the data.  If neither the '-O' nor '-o' options are used,
&lt;prefix&gt; must be present and collate will use it to make an output file name
by appending a suffix depending on the format written (.bam by default).
<p>
If either the -O or -o option is used, &lt;prefix&gt; is optional.  If &lt;prefix&gt;
is absent, collate will write the temporary files to a system-dependent
location (/tmp on UNIX).
<p>
Using -f for fast mode will output <b>only</b> primary alignments that have
either the READ1 <b>or</b> READ2 flags set (but not both).
Any other alignment records will be filtered out.
The collation will only work correctly if there are no more than two reads
for any given QNAME after filtering.
<p>
Fast mode keeps a buffer of alignments in memory so that it can write out
most pairs as soon as they are found instead of storing them in temporary
files.
This allows collate to avoid some work and so finish more quickly compared
to the standard mode.
The number of alignments held can be changed using -r, storing more alignments
uses more memory but increases the number of pairs that can be written early.
<p>
While collate normally randomises the ordering of read pairs, fast mode
does not.
Position-dependent biases that would normally be broken up can remain in the
fast collate output.
It is therefore not a good idea to use fast mode when preparing data for
programs that expect randomly ordered paired reads.
For example using fast collate instead of the standard mode may lead to
significantly different results from aligners that estimate library insert
sizes on batches of reads.
<p>
<b>Options:</b>
<dl><dt><b>-O</b></dt><dd><p>
Output to stdout.  This option cannot be used with '-o'.
</dd><dt><b>-o FILE</b></dt><dd><p>
Write output to FILE.  This option cannot be used with '-O'.
</dd><dt><b>-u</b></dt><dd><p>
Write uncompressed BAM output
</dd><dt><b>-l </b><em>INT</em></dt><dd><p>
Compression level.
[1]
</dd><dt><b>-n </b><em>INT</em></dt><dd><p>
Number of temporary files to use.
[64]
</dd><dt><b>-f</b></dt><dd><p>
Fast mode (primary alignments only).
</dd><dt><b>-r </b><em>INT</em></dt><dd><p>
Number of reads to store in memory (for use with -f).
[10000]
</dd></dl>
<p>
</dd><dt><b>reheader</b></dt><dd><p>
samtools reheader
[<b>-iP</b>]
<em>in.header.sam in.bam</em>
<p>
Replace the header in
<em>in.bam</em>
with the header in
<em>in.header.sam</em>.
This command is much faster than replacing the header with a
BAM&rarr;SAM&rarr;BAM conversion.
<p>
By default this command outputs the BAM or CRAM file to standard
output (stdout), but for CRAM format files it has the option to
perform an in-place edit, both reading and writing to the same file.
No validity checking is performed on the header, nor that it is suitable
to use with the sequence data itself.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-P, --no-PG</b></dt><dd><p>
Do not generate an @PG header line.
</dd><dt><b>-i, --in-place</b></dt><dd><p>
Perform the header edit in-place, if possible.  This only works on CRAM
files and only if there is sufficient room to store the new header.
The amount of space available will differ for each CRAM file.
</dd></dl>
<p>
</dd><dt><b>cat</b></dt><dd><p>
samtools cat [-b list] [-h header.sam] [-o out.bam] &lt;in1.bam&gt; &lt;in2.bam&gt; [ ... ]
<p>
Concatenate BAMs or CRAMs. Although this works on either BAM or CRAM,
all input files must be the same format as each other. The sequence
dictionary of each input file must be identical, although this command
does not check this. This command uses a similar trick to
<b>reheader</b>
which enables fast BAM concatenation.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-b </b><em>FOFN</em></dt><dd><p>
Read the list of input BAM or CRAM files from <em>FOFN</em>.  These are
concatenated prior to any files specified on the command line.
Multiple <b>-b</b> <em>FOFN</em> options may be specified to concatenate
multiple lists of BAM/CRAM files.
</dd><dt><b>-h </b><em>FILE</em></dt><dd><p>
Uses the SAM header from <em>FILE</em>.  By default the header is taken
from the first file to be concatenated.
</dd><dt><b>-o </b><em>FILE</em></dt><dd><p>
Write the concatenated output to <em>FILE</em>.  By default this is sent
to stdout.
</dd></dl>
<p>
</dd><dt><b>rmdup</b></dt><dd><p>
samtools rmdup [-sS] &lt;input.srt.bam&gt; &lt;out.bam&gt;
<p>
<b>This command is obsolete. Use markdup instead.</b>
<p>
Remove potential PCR duplicates: if multiple read pairs have identical
external coordinates, only retain the pair with highest mapping quality.
In the paired-end mode, this command
<b>ONLY</b>
works with FR orientation and requires ISIZE is correctly set. It does
not work for unpaired reads (e.g. two ends mapped to different
chromosomes or orphan reads).
<p>
<b>OPTIONS:</b>
<dl><dt><b>-s</b></dt><dd><p>
Remove duplicates for single-end reads. By default, the command works for
paired-end reads only.
</dd><dt><b>-S</b></dt><dd><p>
Treat paired-end reads and single-end reads.
</dd></dl>
<p>
</dd><dt><b>addreplacerg</b></dt><dd><p>
samtools addreplacerg [-r rg line | -R rg ID] [-m mode] [-l level] [-o out.bam]
&lt;input.bam&gt;
<p>
Adds or replaces read group tags in a file.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-r </b><em>STRING</em></dt><dd><p>
Allows you to specify a read group line to append to the header and applies it
to the reads specified by the -m option. If repeated it automatically adds in
tabs between invocations.
</dd><dt><b>-R </b><em>STRING</em></dt><dd><p>
Allows you to specify the read group ID of an existing @RG line and applies it
to the reads specified.
</dd><dt><b>-m </b><em>MODE</em></dt><dd><p>
If you choose orphan_only then existing RG tags are not overwritten, if you choose
overwrite_all, existing RG tags are overwritten. The default is overwrite_all.
</dd><dt><b>-o </b><em>STRING</em></dt><dd><p>
Write the final output to STRING. The default is to write to stdout.
<p>
By default, samtools tries to select a format based on the output
filename extension; if output is to standard output or no format can be
deduced,
<b>bam</b>
is selected.
</dd></dl>
<p>
</dd><dt><b>calmd</b></dt><dd><p>
samtools calmd [-Eeubr] [-C capQcoef] &lt;aln.bam&gt; &lt;ref.fasta&gt;
<p>
Generate the MD tag. If the MD tag is already present, this command will
give a warning if the MD tag generated is different from the existing
tag. Output SAM by default.
<p>
Calmd can also read and write CRAM files although in most cases it is
pointless as CRAM recalculates MD and NM tags on the fly.  The one
exception to this case is where both input and output CRAM files
have been / are being created with the <em>no_ref</em> option.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-A</b></dt><dd><p>
When used jointly with
<b>-r</b>
this option overwrites the original base quality.
</dd><dt><b>-e</b></dt><dd><p>
Convert a the read base to = if it is identical to the aligned reference
base. Indel caller does not support the = bases at the moment.
</dd><dt><b>-u</b></dt><dd><p>
Output uncompressed BAM
</dd><dt><b>-b</b></dt><dd><p>
Output compressed BAM
</dd><dt><b>-C</b><em> INT</em></dt><dd><p>
Coefficient to cap mapping quality of poorly mapped reads. See the
<b>pileup</b>
command for details. [0]
</dd><dt><b>-r</b></dt><dd><p>
Compute the BQ tag (without -A) or cap base quality by BAQ (with -A).
</dd><dt><b>-E</b></dt><dd><p>
Extended BAQ calculation. This option trades specificity for sensitivity, though the
effect is minor.
</dd></dl>
<p>
</dd><dt><b>targetcut</b></dt><dd><p>
samtools targetcut [-Q minBaseQ] [-i inPenalty] [-0 em0] [-1 em1] [-2 em2] [-f ref] &lt;in.bam&gt;
<p>
This command identifies target regions by examining the continuity of read depth, computes
haploid consensus sequences of targets and outputs a SAM with each sequence corresponding
to a target. When option
<b>-f</b>
is in use, BAQ will be applied. This command is
<b>only</b>
designed for cutting fosmid clones from fosmid pool sequencing [Ref. Kitzman et al. (2010)].
<p>
</dd><dt><b>phase</b></dt><dd><p>
samtools phase [-AF] [-k len] [-b prefix] [-q minLOD] [-Q minBaseQ] &lt;in.bam&gt;
<p>
Call and phase heterozygous SNPs.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-A</b></dt><dd><p>
Drop reads with ambiguous phase.
</dd><dt><b>-b</b><em> STR</em></dt><dd><p>
Prefix of BAM output. When this option is in use, phase-0 reads will be saved in file
<b>STR</b>.0.bam
and phase-1 reads in
<b>STR</b>.1.bam.
Phase unknown reads will be randomly allocated to one of the two files. Chimeric reads
with switch errors will be saved in
<b>STR</b>.chimeric.bam.
[null]
</dd><dt><b>-F</b></dt><dd><p>
Do not attempt to fix chimeric reads.
</dd><dt><b>-k</b><em> INT</em></dt><dd><p>
Maximum length for local phasing. [13]
</dd><dt><b>-q</b><em> INT</em></dt><dd><p>
Minimum Phred-scaled LOD to call a heterozygote. [40]
</dd><dt><b>-Q</b><em> INT</em></dt><dd><p>
Minimum base quality to be used in het calling. [13]
</dd></dl>
<p>
</dd><dt><b>depad</b></dt><dd><p>
samtools depad [-SsCu1] [-T ref.fa] [-o output] &lt;in.bam&gt;
<p>
Converts a BAM aligned against a padded reference to a BAM aligned
against the depadded reference.  The padded reference may contain
verbatim "*" bases in it, but "*" bases are also counted in the
reference numbering.  This means that a sequence base-call aligned
against a reference "*" is considered to be a cigar match ("M" or "X")
operator (if the base-call is "A", "C", "G" or "T").  After depadding
the reference "*" bases are deleted and such aligned sequence
base-calls become insertions.  Similarly transformations apply for
deletions and padding cigar operations.
<p>
<b>OPTIONS:</b>
<dl><dt><b>-S</b></dt><dd><p>
Ignored for compatibility with previous samtools versions.
Previously this option was required if input was in SAM format, but now the
correct format is automatically detected by examining the first few characters
of input.
</dd><dt><b>-s</b></dt><dd><p>
Output in SAM format.  The default is BAM.
</dd><dt><b>-C</b></dt><dd><p>
Output in CRAM format.  The default is BAM.
</dd><dt><b>-u</b></dt><dd><p>
Do not compress the output.  Applies to either BAM or CRAM output
format.
</dd><dt><b>-1</b></dt><dd><p>
Enable fastest compression level.  Only works for BAM or CRAM output.
</dd><dt><b>-T </b><em>FILE</em></dt><dd><p>
Provides the padded reference file.  Note that without this the @SQ
line lengths will be incorrect, so for most use cases this option will
be considered as mandatory.
</dd><dt><b>-o </b><em>FILE</em></dt><dd><p>
Specifies the output filename.  By default output is sent to stdout.
</dd></dl>
<p>
</dd><dt><b>markdup</b></dt><dd><p>
samtools markdup
[<b>-l</b>
<em>length</em>]
[<b>-r</b>]
[<b>-s</b>]
[<b>-T</b>]
[<b>-S</b>]
<em>in.algsort.bam out.bam</em>
<p>
Mark duplicate alignments from a coordinate sorted file that
has been run through fixmate with the -m option.  This program
relies on the MC and ms tags that fixmate provides.
<p>
<b></b>
<dl><dt><b>-l </b><em>INT</em></dt><dd><p>
Expected maximum read length of <em>INT</em> bases.
[300]
</dd><dt><b>-r</b></dt><dd><p>
Remove duplicate reads.
</dd><dt><b>-s</b></dt><dd><p>
Print some basic stats.
</dd><dt><b>-T </b><em>PREFIX</em></dt><dd><p>
Write temporary files to
<em>PREFIX</em><b>.</b><em>samtools</em><b>.</b><em>nnnn</em><b>.</b><em>mmmm</em><b>.</b><em>tmp</em>
</dd><dt><b>-S</b></dt><dd><p>
Mark supplementary reads of duplicates as duplicates.
</dd></dl>
<p>
<pre>
<b>EXAMPLE</b>
<p>
# The first sort can be omitted if the file is already name ordered
samtools sort -n -o namesort.bam example.bam
<p>
# Add ms and MC tags for markdup to use later
samtools fixmate -m namesort.bam fixmate.bam
<p>
# Markdup needs position order
samtools sort -o positionsort.bam fixmate.bam
<p>
# Finally mark duplicates
samtools markdup positionsort.bam markdup.bam
</pre>
</dd><dt><b>help</b>, <b>--help</b></dt><dd><p>
Display a brief usage message listing the samtools commands available.
If the name of a command is also given, e.g.,
<b>samtools help view</b>,
the detailed usage message for that particular command is displayed.
<p>
</dd><dt><b>--version</b></dt><dd><p>
Display the version numbers and copyright information for samtools and
the important libraries used by samtools.
<p>
</dd><dt><b>--version-only</b></dt><dd><p>
Display the full samtools version number in a machine-readable format.
</dd></dl><p>
<h1 id="GLOBAL_OPTIONS"><a href="#GLOBAL_OPTIONS">GLOBAL OPTIONS</a></h1>
<p>
Several long-options are shared between multiple samtools subcommands:
<b>--input-fmt</b>, <b>--input-fmt-option</b>, <b>--output-fmt</b>,
<b>--output-fmt-option</b>, and <b>--reference</b>.
The input format is typically auto-detected so specifying the format
is usually unnecessary and the option is included for completeness.
Note that not all subcommands have all options.  Consult the subcommand
help for more details.
<p>
Format strings recognised are "sam", "bam" and "cram".  They may be
followed by a comma separated list of options as <em>key</em> or
<em>key</em>=<em>value</em>. See below for examples.
<p>
The <b>fmt-option</b> arguments accept either a single <em>option</em> or
<em>option</em>=<em>value</em>.  Note that some options only work on some
file formats and only on read or write streams.  If value is
unspecified for a boolean option, the value is assumed to be 1.  The
valid options are as follows.
<dl><dt><b>level=</b><em>INT</em></dt><dd><p>
Output only. Specifies the compression level from 1 to 9, or 0 for
uncompressed.
</dd><dt><b>nthreads=</b><em>INT</em></dt><dd><p>
Specifies the number of threads to use during encoding and/or
decoding.  For BAM this will be encoding only.  In CRAM the threads
are dynamically shared between encoder and decoder.
</dd><dt><b>reference=</b><em>fasta_file</em></dt><dd><p>
Specifies a FASTA reference file for use in CRAM encoding or decoding.
It usually is not required for decoding except in the situation of the
MD5 not being obtainable via the REF_PATH or REF_CACHE environment variables.
</dd><dt><b>decode_md=</b><em>0|1</em></dt><dd><p>
CRAM input only; defaults to 1 (on).  CRAM does not typically store
MD and NM tags, preferring to generate them on the fly.  This option
controls this behaviour.  It can be particularly useful when combined
with a file encoded using store_md=1 and store_nm=1.
</dd><dt><b>store_md=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  CRAM normally only stores MD
tags when no reference is unknown and lets the decoder generate these
values on-the-fly (see decode_md).
</dd><dt><b>store_nm=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  CRAM normally only stores NM
tags when no reference is unknown and lets the decoder generate these
values on-the-fly (see decode_md).
</dd><dt><b>ignore_md5=</b><em>0|1</em></dt><dd><p>
CRAM input only; defaults to 0 (off).  When enabled, md5 checksum
errors on the reference sequence and block checksum errors within CRAM
are ignored.  Use of this option is strongly discouraged.
</dd><dt><b>required_fields=</b><em>bit-field</em></dt><dd><p>
CRAM input only; specifies which SAM columns need to be populated.
By default all fields are used.  Limiting the decode to specific
columns can have significant performance gains.  The bit-field is a
numerical value constructed from the following table.
<table align="center">
<tr><td>0x1</td><td>SAM_QNAME</td></tr>
<tr><td>0x2</td><td>SAM_FLAG</td></tr>
<tr><td>0x4</td><td>SAM_RNAME</td></tr>
<tr><td>0x8</td><td>SAM_POS</td></tr>
<tr><td>0x10</td><td>SAM_MAPQ</td></tr>
<tr><td>0x20</td><td>SAM_CIGAR</td></tr>
<tr><td>0x40</td><td>SAM_RNEXT</td></tr>
<tr><td>0x80</td><td>SAM_PNEXT</td></tr>
<tr><td>0x100</td><td>SAM_TLEN</td></tr>
<tr><td>0x200</td><td>SAM_SEQ</td></tr>
<tr><td>0x400</td><td>SAM_QUAL</td></tr>
<tr><td>0x800</td><td>SAM_AUX</td></tr>
<tr><td>0x1000</td><td>SAM_RGAUX</td></tr>
</table>
</dd><dt><b>name_prefix=</b><em>string</em></dt><dd><p>
CRAM input only; defaults to output filename.  Any sequences with
auto-generated read names will use <em>string</em> as the name prefix.
</dd><dt><b>multi_seq_per_slice=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  By default CRAM generates one
container per reference sequence, except in the case of many small
references (such as a fragmented assembly).
</dd><dt><b>version=</b><em>major.minor</em></dt><dd><p>
CRAM output only.  Specifies the CRAM version number.  Acceptable
values are "2.1" and "3.0".
</dd><dt><b>seqs_per_slice=</b><em>INT</em></dt><dd><p>
CRAM output only; defaults to 10000.
</dd><dt><b>slices_per_container=</b><em>INT</em></dt><dd><p>
CRAM output only; defaults to 1.  The effect of having multiple slices
per container is to share the compression header block between
multiple slices.  This is unlikely to have any significant impact
unless the number of sequences per slice is reduced.  (Together these
two options control the granularity of random access.)
</dd><dt><b>embed_ref=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  If 1, this will store portions
of the reference sequence in each slice, permitting decode without
having requiring an external copy of the reference sequence.
</dd><dt><b>no_ref=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  If 1, sequences will be stored
verbatim with no reference encoding.  This can be useful if no
reference is available for the file.
</dd><dt><b>use_bzip2=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  Permits use of bzip2 in CRAM
block compression.
</dd><dt><b>use_lzma=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  Permits use of lzma in CRAM
block compression.
</dd><dt><b>lossy_names=</b><em>0|1</em></dt><dd><p>
CRAM output only; defaults to 0 (off).  If 1, templates with all
members within the same CRAM slice will have their read names
removed.  New names will be automatically generated during decoding.
Also see the <b>name_prefix</b> option.
</dd></dl>
<p>
For example:
<pre>
samtools view --input-fmt-option decode_md=0
    --output-fmt cram,version=3.0 --output-fmt-option embed_ref
    --output-fmt-option seqs_per_slice=2000 -o foo.cram foo.bam
</pre>
<p>
<h1 id="REFERENCE_SEQUENCES"><a href="#REFERENCE_SEQUENCES">REFERENCE SEQUENCES</a></h1>
<p>
The CRAM format requires use of a reference sequence for both reading
and writing.
<p>
When reading a CRAM the <b>@SQ</b> headers are interrogated to identify
the reference sequence MD5sum (<b>M5:</b> tag) and the local reference
sequence filename (<b>UR:</b> tag).  Note that <em>http://</em> and
<em>ftp://</em> based URLs in the UR: field are not used, but local fasta
filenames (with or without <em>file://</em>) can be used.
<p>
To create a CRAM the <b>@SQ</b> headers will also be read to identify
the reference sequences, but M5: and UR: tags may not be present. In
this case the <b>-T</b> and <b>-t</b> options of samtools view may be
used to specify the fasta or fasta.fai filenames respectively
(provided the .fasta.fai file is also backed up by a .fasta file).
<p>
The search order to obtain a reference is:
<p>
Use any local file specified by the command line options (eg -T).
<p>
Look for MD5 via REF_CACHE environment variable.
<p>
Look for MD5 in each element of the REF_PATH environment variable.
<p>
Look for a local file listed in the UR: header tag.
<p>
<h1 id="ENVIRONMENT_VARIABLES"><a href="#ENVIRONMENT_VARIABLES">ENVIRONMENT VARIABLES</a></h1>
<p>
<dl><dt><b>HTS_PATH</b></dt><dd><p>
A colon-separated list of directories in which to search for HTSlib plugins.
If $HTS_PATH starts or ends with a colon or contains a double colon (<b>::</b>),
the built-in list of directories is searched at that point in the search.
<p>
If no HTS_PATH variable is defined, the built-in list of directories
specified when HTSlib was built is used, which typically includes
<b>/usr/local/libexec/htslib</b> and similar directories.
<p>
</dd><dt><b>REF_PATH</b></dt><dd><p>
A colon separated (semi-colon on Windows) list of locations in which
to look for sequences identified by their MD5sums.  This can be either
a list of directories or URLs. Note that if a URL is included then the
colon in http:// and ftp:// and the optional port number will be
treated as part of the URL and not a PATH field separator.
For URLs, the text <b>%s</b> will be replaced by the MD5sum being
read.
<p>
If no REF_PATH has been specified it will default to
<b>http://www.ebi.ac.uk/ena/cram/md5/%s</b> and if REF_CACHE is also unset,
it will be set to <b>$XDG_CACHE_HOME/hts-ref/%2s/%2s/%s</b>.
If <b>$XDG_CACHE_HOME</b> is unset, <b>$HOME/.cache</b> (or a local system
temporary directory if no home directory is found) will be used similarly.
<p>
</dd><dt><b>REF_CACHE</b></dt><dd><p>
This can be defined to a single directory housing a local cache of
references.  Upon downloading a reference it will be stored in the
location pointed to by REF_CACHE.  When reading a reference it will be
looked for in this directory before searching REF_PATH.  To avoid many
files being stored in the same directory, a pathname may be
constructed using %<em>num</em>s and %s notation, consuming <em>num</em>
characters of the MD5sum.  For example
<b>/local/ref_cache/%2s/%2s/%s</b> will create 2 nested subdirectories
with the filenames in the deepest directory being the last 28
characters of the md5sum.
<p>
The REF_CACHE directory will be searched for before attempting to load
via the REF_PATH search list.  If no REF_PATH is defined, both
REF_PATH and REF_CACHE will be automatically set (see above), but if
REF_PATH is defined and REF_CACHE not then no local cache is used.
<p>
To aid population of the REF_CACHE directory a script
<b>misc/seq_cache_populate.pl</b> is provided in the Samtools
distribution. This takes a fasta file or a directory of fasta files
and generates the MD5sum named files.
</dd></dl><p>
<h1 id="EXAMPLES"><a href="#EXAMPLES">EXAMPLES</a></h1>
<ul><li>Import SAM to BAM when
<b>@SQ</b>
lines are present in the header:
<pre>
samtools view -bS aln.sam &gt; aln.bam
</pre>
If
<b>@SQ</b>
lines are absent:
<pre>
samtools faidx ref.fa
samtools view -bt ref.fa.fai aln.sam &gt; aln.bam
</pre>
where
<em>ref.fa.fai</em>
is generated automatically by the
<b>faidx</b>
command.
<p>
</li>
<li>Convert a BAM file to a CRAM file using a local reference sequence.
<pre>
samtools view -C -T ref.fa aln.bam &gt; aln.cram
</pre>
</li>
<li>Attach the
<b>RG</b>
tag while merging sorted alignments:
<pre>
perl -e 'print "@RG\\tID:ga\\tSM:hs\\tLB:ga\\tPL:Illumina\\n@RG\\tID:454\\tSM:hs\\tLB:454\\tPL:454\\n"' &gt; rg.txt
samtools merge -rh rg.txt merged.bam ga.bam 454.bam
</pre>
The value in a
<b>RG</b>
tag is determined by the file name the read is coming from. In this
example, in the
<em>merged.bam</em>,
reads from
<em>ga.bam</em>
will be attached
<em>RG:Z:ga</em>,
while reads from
<em>454.bam</em>
will be attached
<em>RG:Z:454</em>.
<p>
</li>
<li>Convert a BAM file to a CRAM with NM and MD tags stored verbatim
rather than calculating on the fly during CRAM decode, so that mixed
data sets with MD/NM only on some records, or NM calculated using
different definitions of mismatch, can be decoded without change.  The
second command demonstrates how to decode such a file.  The request to
not decode MD here is turning off auto-generation of both MD and NM;
it will still emit the MD/NM tags on records that had these stored
verbatim.
<pre>
samtools view -C --output-fmt-option store_md=1 --output-fmt-option store_nm=1 -o aln.cram aln.bam
samtools view --input-fmt-option decode_md=0 -o aln.new.bam aln.cram
</pre>
</li>
<li>An alternative way of achieving the above is listing multiple options
after the <b>--output-fmt</b> or <b>-O</b> option.  The commands below
are equivalent to the two above.
<pre>
samtools view -O cram,store_md=1,store_nm=1 -o aln.cram aln.bam
samtools view --input-fmt cram,decode_md=0 -o aln.new.bam aln.cram
</pre>
<p>
</li>
<li>Call SNPs and short INDELs:
<pre>
samtools mpileup -uf ref.fa aln.bam | bcftools call -mv &gt; var.raw.vcf
bcftools filter -s LowQual -e '%QUAL&lt;20 || DP&gt;100' var.raw.vcf  &gt; var.flt.vcf
</pre>
The
<b>bcftools filter</b>
command marks low quality sites and sites with the read depth exceeding
a limit, which should be adjusted to about twice the average read depth
(bigger read depths usually indicate problematic regions which are
often enriched for artefacts).  One may consider to add
<b>-C50</b>
to
<b>mpileup</b>
if mapping quality is overestimated for reads containing excessive
mismatches. Applying this option usually helps
<b>BWA-short</b>
but may not other mappers.
<p>
Individuals are identified from the
<b>SM</b>
tags in the
<b>@RG</b>
header lines. Individuals can be pooled in one alignment file; one
individual can also be separated into multiple files. The
<b>-P</b>
option specifies that indel candidates should be collected only from
read groups with the
<b>@RG-PL</b>
tag set to
<em>ILLUMINA</em>.
Collecting indel candidates from reads sequenced by an indel-prone
technology may affect the performance of indel calling.
<p>
</li>
<li>Generate the consensus sequence for one diploid individual:
<pre>
samtools mpileup -uf ref.fa aln.bam | bcftools call -c | vcfutils.pl vcf2fq &gt; cns.fq
</pre>
</li>
<li>Phase one individual:
<pre>
samtools calmd -AEur aln.bam ref.fa | samtools phase -b prefix - &gt; phase.out
</pre>
The
<b>calmd</b>
command is used to reduce false heterozygotes around INDELs.
<p>
<p>
</li>
<li>Dump BAQ applied alignment for other SNP callers:
<pre>
samtools calmd -bAr aln.bam &gt; aln.baq.bam
</pre>
It adds and corrects the
<b>NM</b>
and
<b>MD</b>
tags at the same time. The
<b>calmd</b>
command also comes with the
<b>-C</b>
option, the same as the one in
<b>pileup</b>
and
<b>mpileup</b>.
Apply if it helps.
<p>
</ul>
<h1 id="LIMITATIONS"><a href="#LIMITATIONS">LIMITATIONS</a></h1>
<p>
<ul><li>Unaligned words used in bam_import.c, bam_endian.h, bam.c and bam_aux.c.
</li>
<li>Samtools paired-end rmdup does not work for unpaired reads (e.g. orphan
reads or ends mapped to different chromosomes). If this is a concern,
please use Picard's MarkDuplicates which correctly handles these cases,
although a little slower.
<p>
</ul>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Heng Li from the Sanger Institute wrote the original C version of samtools.
Bob Handsaker from the Broad Institute implemented the BGZF library.
James Bonfield from the Sanger Institute developed the CRAM implementation.
John Marshall and Petr Danecek contribute to the source code and various
people from the 1000 Genomes Project have contributed to the SAM format
specification.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<em>bcftools</em>(1),
<em>sam</em>(5),
<em>tabix</em>(1)
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
<br>
File format specification of SAM/BAM,CRAM,VCF/BCF: &lt;<a href="http://samtools.github.io/hts-specs">http://samtools.github.io/hts-specs</a>&gt;
<br>
Samtools latest source: &lt;<a href="https://github.com/samtools/samtools">https://github.com/samtools/samtools</a>&gt;
<br>
HTSlib latest source: &lt;<a href="https://github.com/samtools/htslib">https://github.com/samtools/htslib</a>&gt;
<br>
Bcftools website: &lt;<a href="http://samtools.github.io/bcftools">http://samtools.github.io/bcftools</a>&gt;
