---
permalink: /doc/1.18/samtools-samples.html
layout: manpage
title: samtools-samples(1) manual page
package: samtools-1.18
date: 25 July 2023
section_text: Bioinformatics tools
redirect_from: /doc/1.18/samtools-samples.1.html
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools samples &ndash; prints the samples from an alignment file
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
<b>samtools samples</b>
[<em>options</em>]
<em>(&lt;input&gt;|stdin)</em>
<p>
<b>samtools samples</b>
[<em>options</em>]
<b>-X</b>
<em>f1.bam</em> <em>f2.bam</em> ... <em>f1.bam.bai</em> <em>f2.bam.bai</em> ... 
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
Print the sample names found in the read-groups and the path to the reference
genome from alignment files. The output of this tool can be used to create an
input for any popular workflow manager. The input is a list of SAM/BAM/CRAM
files, or the path to those files can be provided via stdin. The output is
tab-delimited containing the sample name as the first column, the path to the
alignment file as the second column, the path to the reference genome as a
third optional column and a single character flag (Y/N) indicating whether the
alignment file is indexed or not as a fourth optional column.
If no reference is found for an alignment, a dot (.) will be used in the
reference path column. If no sample is available in any read-group header, a
dot (.) will be used as the sample name. If a BAM file contains more than one
sample, one line will be printed for each sample.
<p>
<h1 id="OPTIONS"><a href="#OPTIONS">OPTIONS</a></h1>
<dl><dt><b>-?</b></dt><dd><p>
print help and exit
</dd><dt><b>-h</b></dt><dd><p>
print a header
</dd><dt><b>-i</b></dt><dd><p>
test if the file is indexed. Add an extra column to the output with a single
character value (Y/N).
</dd><dt><b>-T </b><em>TAG</em></dt><dd><p>
provide the sample tag name from the @RG line [SM].
</dd><dt><b>-o </b><em>FILE</em></dt><dd><p>
output file [stdout].
</dd><dt><b>-f </b><em>FILE</em></dt><dd><p>
load an indexed fasta file in the collection of references. Can be used multiple
times. Add an extra column with the path to the reference file.
</dd><dt><b>-F </b><em>FILE</em></dt><dd><p>
read a file containing the paths to indexed fasta files. One path per line.
</dd><dt><b>-X</b></dt><dd><p>
use a custom index file.
<p>
</dd></dl>
<h1 id="EXAMPLES"><a href="#EXAMPLES">EXAMPLES</a></h1>
<ul><li>print the samples from a set of BAM/SAM files, with a header. There is no sample
defined in the header of 'example.sam', so a dot is used for the sample name.
<pre>
$ samtools  samples -h S*.bam *.sam
#SM	PATH
S1	S1.bam
S2	S2.bam
S3	S3.bam
S4	S4.bam
S5	S5.bam
.	example.sam
</pre>
</li>
<li>print the samples from a set of BAM/SAM files, with a header, print whether the
file is indexed.
<pre>
$  samtools  samples -i -h S*.bam *.sam
#SM	PATH	INDEX
S1	S1.bam	Y
S2	S2.bam	Y
S3	S3.bam	Y
S4	S4.bam	Y
S5	S5.bam	Y
.	example.sam	N
</pre>
</li>
<li>print whether the files are indexed using custom bai files.
<pre>
$ samtools samples -i -h -X S1.bam S2.bam S1.bam.bai S2.bam.bai
#SM	PATH	INDEX
S1	S1.bam	Y
S2	S2.bam	Y
</pre>
</li>
<li>read a tab delimited input &lt;file&gt;(tab)&lt;bai&gt; and print whether the files are
indexed using custom bai files.
<pre>
$ find . -type f \( -name "S*.bam" -o -name "S*.bai" \) | sort | paste - - | samtools samples -i -h -X
#SM	PATH	INDEX
S1	./S1.bam	Y
S2	./S2.bam	Y
S3	./S3.bam	Y
S4	./S4.bam	Y
S5	./S5.bam	Y
</pre>
</li>
<li>print the samples from a set of BAM/CRAM files, with a header, use '@RG/LB'
instead of '@RG/SM'.
<pre>
$ samtools  samples -h -T LB S*.bam
#LB	PATH
S1	S1.bam
S2	S2.bam
S3	S3.bam
S4	S4.bam
S5Lib1	S5.bam
S5Lib2	S5.bam
</pre>
</li>
<li>pipe a list of BAM/CRAM files , pipe it into  samtools samples.
<pre>
$ find . -type f \( -name "S*.bam" -o -name "*.cram" \) | samtools  samples -h
#SM	PATH
S5	./S5.bam
S2	./S2.bam
S4	./S4.bam
S3	./S3.bam
S1	./example.cram
S1	./S1.bam
</pre>
</li>
<li>provide two reference sequences with option '-f', print the associated reference
for each BAM files.
<pre>
$ samtools  samples  -h -f reference.fa -f example.fa S*.bam *.sam *.cram
#SM	PATH	REFERENCE
S1	S1.bam	reference.fa
S2	S2.bam	reference.fa
S3	S3.bam	reference.fa
S4	S4.bam	reference.fa
S5	S5.bam	reference.fa
.	example.sam	example.fa
S1	example.cram	example.fa
</pre>
</li>
<li>provide a list of reference sequences with option '-F', print the associated
reference for each BAM files.
<pre>
$ cat references.list
reference.fa
example.fa
$ samtools  samples  -h -F references.list S*.bam *.sam *.cram
#SM	PATH	REFERENCE
S1	S1.bam	reference.fa
S2	S2.bam	reference.fa
S3	S3.bam	reference.fa
S4	S4.bam	reference.fa
S5	S5.bam	reference.fa
.	example.sam	example.fa
S1	example.cram	example.fa
</pre>
<p>
</li></ul>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Written by Pierre Lindenbaum from Institut du Thorax U1087, Nantes, France.
<p>
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
