---
permalink: /doc/1.19/samtools-fasta.html
layout: manpage
title: samtools-fasta(1) manual page
package: samtools-1.19
date: 12 December 2023
section_text: Bioinformatics tools
redirect_from: /doc/1.19/samtools-fasta.1.html
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools-fasta, samtools-fastq &ndash; converts a SAM/BAM/CRAM file to FASTA or FASTQ
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
samtools <a href="samtools-fastq.html">fastq</a>
[<em>options</em>]
<em>in.bam</em>
<br>
samtools <a href="samtools-fasta.html">fasta</a>
[<em>options</em>]
<em>in.bam</em>
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
Converts a BAM or CRAM into either FASTQ or FASTA format depending on the
command invoked. The files will be automatically compressed if the
file names have a .gz, .bgz, or .bgzf extension.
<p>
Note this command is attempting to reverse the alignment process, so
if the aligner took a single input FASTQ and produced multiple SAM
records via supplementary and/or secondary alignments, then converting
back to FASTQ again should produce the original single FASTA / FASTQ
record.  By default it will not attempt to records for supplementary
and secondary alignments, but see the <b>-F</b> option for more details.
<p>
If the input contains read-pairs which are to be interleaved or
written to separate files in the same order, then the input should
be first collated by name.
Use
<b>samtools collate</b>
or
<b>samtools sort -n</b>
to ensure this.
<p>
For each different QNAME, the input records are categorised according to
the state of the READ1 and READ2 flag bits.
The three categories used are:
<p>
1 : Only READ1 is set.
<p>
2 : Only READ2 is set.
<p>
0 : Either both READ1 and READ2 are set; or neither is set.
<p>
The exact meaning of these categories depends on the sequencing technology
used.
It is expected that ordinary single and paired-end sequencing reads will be
in categories 1 and 2 (in the case of paired-end reads, one read of the pair
will be in category 1, the other in category 2).
Category 0 is essentially a &ldquo;catch-all&rdquo; for reads that do not
fit into a simple paired-end sequencing model.
<p>
For each category only one sequence will be written for a given QNAME.
If more than one record is available for a given QNAME and category,
the first in input file order that has quality values will be used.
If none of the candidate records has quality values, then the first in
input file order will be used instead.
<p>
Sequences will be written to standard output unless one of the
<b>-1</b>, <b>-2</b>, <b>-o</b>, or <b>-0</b>
options is used, in which case sequences for that category will be written to
the specified file.
The same filename may be specified with multiple options, in which case the
sequences will be multiplexed in order of occurrence.
<p>
If a singleton file is specified using the
<b>-s</b>
option then only paired sequences will be output for categories 1 and 2;
paired meaning that for a given QNAME there are sequences for both
category 1
<b>and</b>
2.
If there is a sequence for only one of categories 1 or 2 then it will be
diverted into the specified singletons file.
This can be used to prepare fastq files for programs that cannot handle
a mixture of paired and singleton reads.
<p>
The
<b>-s</b>
option only affects category 1 and 2 records.
The output for category 0 will be the same irrespective of the use of this
option.
<p>
The sequence generated will be for the entire sequence recorded in the
SAM record (and quality if appropriate).  This means if it has
soft-clipped CIGAR records then the soft-clipped data will be in the
output FASTA/FASTQ.  Hard-clipped data is, by definition, absent from
the SAM record and hence will be absent in any FASTA/FASTQ produced.
<p>
The filter options order of precedence is -d, -f, -F, --rf and -G.
<p>
<h1 id="OPTIONS"><a href="#OPTIONS">OPTIONS</a></h1>
<dl><dt><b>-n</b></dt><dd><p>
By default, either '/1' or '/2' is added to the end of read names
where the corresponding READ1 or READ2 FLAG bit is set.
Using
<b>-n</b>
causes read names to be left as they are.
</dd><dt><b>-N</b></dt><dd><p>
Always add either '/1' or '/2' to the end of read names
even when put into different files.
</dd><dt><b>-O</b></dt><dd><p>
Use quality values from OQ tags in preference to standard quality string
if available.
</dd><dt><b>-s FILE</b></dt><dd><p>
Write singleton reads to FILE.
</dd><dt><b>-t</b></dt><dd><p>
Copy RG, BC and QT tags to the FASTQ header line, if they exist.
</dd><dt><b>-T TAGLIST</b></dt><dd><p>
Specify a comma-separated list of tags to copy to the FASTQ header line, if
they exist.
<b>TAGLIST</b> can be blank or <b>*</b> to indicate all tags should be copied to
the output.
If using <b>*</b>, be careful to quote it to avoid unwanted shell expansion.
</dd><dt><b>-1 FILE</b></dt><dd><p>
Write reads with the READ1 FLAG set (and READ2 not set) to FILE instead of
outputting them.
If the
<b>-s</b>
option is used, only paired reads will be written to this file.
</dd><dt><b>-2 FILE</b></dt><dd><p>
Write reads with the READ2 FLAG set (and READ1 not set) to FILE instead of
outputting them.
If the
<b>-s</b>
option is used, only paired reads will be written to this file.
</dd><dt><b>-o FILE</b></dt><dd><p>
Write reads with either READ1 FLAG or READ2 flag set to FILE instead
of outputting them to stdout.  This is equivalent to <b>-1</b> FILE
<b>-2</b> FILE.
</dd><dt><b>-0 FILE</b></dt><dd><p>
Write reads where the READ1 and READ2 FLAG bits set are either both set
or both unset to FILE instead of outputting them.
</dd><dt><b>-f </b><em>INT</em></dt><dd><p>
Only output alignments with all bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-F </b><em>INT</em><b>, </b><em>,</em><b>--excl-flags</b><em> INT </em><b>,</b><em>--exclude-flags</em><b> INT </b></dt><dd><p>
Do not output alignments with any bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0x900].
This defaults to 0x900 representing filtering of secondary and
supplementary alignments.
</dd><dt><b>--rf </b><em>INT</em><b> , --incl-flags </b><em>INT</em><b>, --include-flags </b><em>INT</em></dt><dd><p>
Only output alignments with any bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/),
in octal by beginning with `0' (i.e. /^0[0-7]+/), as a decimal number
not beginning with '0' or as a comma-separated list of flag names [0].
</dd><dt><b>-G </b><em>INT</em></dt><dd><p>
Only EXCLUDE reads with all of the bits set in
<em>INT</em>
present in the FLAG field.
<em>INT</em>
can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/)
or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0].
</dd><dt><b>-d </b><em>TAG</em><b>[:</b><em>VAL</em><b>]</b></dt><dd><p>
Only output alignments containing an auxiliary tag matching both
<em>TAG</em> and <em>VAL</em>.  If <em>VAL</em> is omitted then any value is
accepted.  The tag types supported are i, f, Z, A and H.  "B" arrays
are not supported.  This is comparable to the method used in
<b>samtools view -d</b>, but for single values only (i.e. there is no
sibling <b>-D</b> option).
</dd><dt><b>-i</b></dt><dd><p>
add Illumina Casava 1.8 format entry to header (eg 1:N:0:ATCACG)
</dd><dt><b>-c [0..9]</b></dt><dd><p>
set compression level when writing gz or bgzf fastq files.
</dd><dt><b>--i1 FILE</b></dt><dd><p>
write first index reads to FILE
</dd><dt><b>--i2 FILE</b></dt><dd><p>
write second index reads to FILE
</dd><dt><b>--barcode-tag TAG</b></dt><dd><p>
aux tag to find index reads in [default: BC]
</dd><dt><b>--quality-tag TAG</b></dt><dd><p>
aux tag to find index quality in [default: QT]
</dd><dt><b>-@, --threads </b><em>INT</em></dt><dd><p>
Number of input/output compression threads to use in addition to main thread [0].
</dd><dt><b>--index-format STR</b></dt><dd><p>
string to describe how to parse the barcode and quality tags. For example:
<p>
<dl><dt><b>i14i8</b></dt><dd><p>
the first 14 characters are index 1, the next 8 characters are index 2
</dd><dt><b>n8i14</b></dt><dd><p>
ignore the first 8 characters, and use the next 14 characters for index 1
<p>
If the tag contains a separator, then the numeric part can be replaced with '*' to
mean 'read until the separator or end of tag', for example:
</dd><dt><b>n*i*</b></dt><dd><p>
ignore the left part of the tag until the separator, then use the second part
</dd></dl>
<p>
</dd></dl>
<h1 id="EXAMPLES"><a href="#EXAMPLES">EXAMPLES</a></h1>
Starting from a coordinate sorted file, output paired reads to
separate files, discarding singletons, supplementary and secondary reads.
The resulting files can be used with, for example, the
<b>bwa</b>
aligner.
<pre>
samtools collate -u -O in_pos.bam | \
samtools fastq -1 paired1.fq -2 paired2.fq -0 /dev/null -s /dev/null -n
</pre>
<p>
Starting with a name collated file, output paired and singleton reads
in a single file, discarding supplementary and secondary reads.
To get all of the reads in a single file, it is necessary to redirect the
output of samtools fastq.
The output file is suitable for use with
<b>bwa mem -p</b>
which understands interleaved files containing a mixture of paired and
singleton reads.
<pre>
samtools fastq -0 /dev/null in_name.bam &gt; all_reads.fq
</pre>
<p>
Output paired reads in a single file, discarding supplementary and
secondary reads.
Save any singletons in a separate file.
Append /1 and /2 to read names.
This format is suitable for use by
<b>NextGenMap</b>
when using its
<b>-p</b> and <b>-q</b> options.
With this aligner, paired reads must be mapped separately to the singletons.
<pre>
samtools fastq -0 /dev/null -s single.fq -N in_name.bam &gt; paired.fq
</pre>
<p>
<h1 id="BUGS"><a href="#BUGS">BUGS</a></h1>
<ul><li>The way of specifying output files is far too complicated and easy to get wrong.
<p>
</li></ul>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Written by Heng Li, with modifications by Martin Pollard and Jennifer Liddle,
all from the Sanger Institute.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<a href="samtools.html"><em>samtools</em></a> (1),
<a href="samtools-faidx.html"><em>samtools-faidx</em></a> (1),
<a href="samtools-fqidx.html"><em>samtools-fqidx</em></a> (1)
<a href="samtools-import.html"><em>samtools-import</em></a> (1)
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
