---
permalink: /doc/1.19/samtools-tview.html
layout: manpage
title: samtools-tview(1) manual page
package: samtools-1.19
date: 12 December 2023
section_text: Bioinformatics tools
redirect_from: /doc/1.19/samtools-tview.1.html
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
samtools tview &ndash; display alignments in a curses-based interactive viewer.
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
<b>samtools tview</b>
[<b>-p</b>
<em>chr:pos</em>]
[<b>-s</b>
<em>STR</em>]
[<b>-d</b>
<em>display</em>]
<em>in.sorted.bam</em>
[<em>ref.fasta</em>]
	
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
Text alignment viewer (based on the ncurses library). In the viewer,
press `?' for help and press `g' to check the alignment start from a
region in the format like `chr10:10,000,000' or `=10,000,000' when
viewing the same reference sequence.
<p>
The top line shows the reference sequence, or '<b>N</b>'s if unknown.
Underneath this is the consensus, derived from the sequence
alignments.  Below the consensus the sequence alignment records are
shown.  Uppercase and lowercase is used to distinguish the sequence
strand, with uppercase being the top/forward strand.
<p>
When the reference is known, both consensus and alignment record
sequences are displayed in a dot-notation where a matching character
is shown as '<b>.</b>'  (forward strand) or '<b>,</b>' (reverse strand) and only
mismatching bases and missing bases are shown.  This mode can be
toggled with the "."  command.
<p>
<h1 id="OPTIONS"><a href="#OPTIONS">OPTIONS</a></h1>
<dl><dt><b>-d</b><em> display</em></dt><dd><p>
Output as (H)tml, (C)urses or (T)ext.
<p>
The width of generated text is controlled by the COLUMNS environment
variable or the <b>-w</b> option for non-curses outputs.
Note this may be a local shell variable so it may need
exporting first or specifying on the command line prior to the
command.  For example
<pre>
export COLUMNS ; samtools tview -d T -p 1:234567 in.bam
</pre>
</dd><dt><b>-p</b><em> chr:pos</em></dt><dd><p>
Go directly to this position
</dd><dt><b>-s</b><em> STR</em></dt><dd><p>
Display only alignments from this sample or read group.  <b>STR</b>
must match either an <b>ID</b> or <b>SM</b> field in an <b>@RG</b> header
record.  For example
<pre>
samtools tview -p chr20:10M -s NA12878 grch38.fa
</pre>
</dd><dt><b>-w</b><em> INT</em></dt><dd><p>
Specifies the display width when using the HTML or Text output modes.
</dd><dt><b>-X</b></dt><dd><p>
If this option is set, it will allows user to specify customized index file location(s) if the data
folder does not contain any index file. Example usage: samtools tview [options] -X &lt;/data_folder/data.bam&gt; [/index_folder/index.bai] [ref.fasta]
<p>
</dd></dl>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Written by Heng Li from the Sanger Institute.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<a href="samtools.html"><em>samtools</em></a> (1)
<p>
Samtools website: &lt;<a href="http://www.htslib.org/">http://www.htslib.org/</a>&gt;
