---
permalink: /doc/1.19/tabix.html
layout: manpage
title: tabix(1) manual page
package: htslib-1.19
date: 12 December 2023
section_text: Bioinformatics tools
redirect_from: /doc/1.19/tabix.1.html
---
<h1 id="NAME"><a href="#NAME">NAME</a></h1>
<p>
tabix &ndash; Generic indexer for TAB-delimited genome position files
<h1 id="SYNOPSIS"><a href="#SYNOPSIS">SYNOPSIS</a></h1>
<p>
<b>tabix</b>
[<b>-0lf</b>]
[<b>-p</b>
gff|bed|sam|vcf]
[<b>-s</b>
<em>seqCol</em>]
[<b>-b</b>
<em>begCol</em>]
[<b>-e</b>
<em>endCol</em>]
[<b>-S</b>
<em>lineSkip</em>]
[<b>-c</b>
<em>metaChar</em>]
<em>in.tab.bgz</em>
[<em>region1 </em>[<em>region2 </em>[<em>...</em>]]]
<p>
<h1 id="DESCRIPTION"><a href="#DESCRIPTION">DESCRIPTION</a></h1>
<p>
Tabix indexes a TAB-delimited genome position file
<em>in.tab.bgz</em>
and creates an index file
(<em>in.tab.bgz.tbi</em>
or 
<em>in.tab.bgz.csi</em>)
when
<em>region</em>
is absent from the command-line. The input data file must be position
sorted and compressed by
<b>bgzip</b>
which has a
<b>gzip</b>(1)
like interface.
<p>
After indexing, tabix is able to quickly retrieve data lines overlapping
<em>regions</em>
specified in the format "chr:beginPos-endPos".
(Coordinates specified in this region format are 1-based and inclusive.)
<p>
Fast data retrieval also
works over network if URI is given as a file name and in this case the
index file will be downloaded if it is not present locally.
<p>
The tabix
(<em>.tbi</em>)
and BAI index formats can handle individual chromosomes up to 512 Mbp
(2^29 bases) in length.
If your input file might contain data lines with begin or end positions
greater than that, you will need to use a CSI index.
<p>
<h1 id="INDEXING_OPTIONS"><a href="#INDEXING_OPTIONS">INDEXING OPTIONS</a></h1>
<dl><dt><b>-0, --zero-based</b></dt><dd><p>
Specify that the position in the data file is 0-based half-open
(e.g. UCSC files) rather than 1-based.
</dd><dt><b>-b, --begin </b><em>INT</em></dt><dd><p>
Column of start chromosomal position. [4]
</dd><dt><b>-c, --comment </b><em>CHAR</em></dt><dd><p>
Skip lines started with character CHAR. [#]
</dd><dt><b>-C, --csi</b></dt><dd><p>
Produce CSI format index instead of classical tabix or BAI style indices.
</dd><dt><b>-e, --end </b><em>INT</em></dt><dd><p>
Column of end chromosomal position. The end column can be the same as the
start column. [5]
</dd><dt><b>-f, --force </b></dt><dd><p>
Force to overwrite the index file if it is present.
</dd><dt><b>-m, --min-shift </b><em>INT</em></dt><dd><p>
Set minimal interval size for CSI indices to 2^INT [14]
</dd><dt><b>-p, --preset </b><em>STR</em></dt><dd><p>
Input format for indexing. Valid values are: gff, bed, sam, vcf.
This option should not be applied together with any of
<b>-s</b>, <b>-b</b>, <b>-e</b>, <b>-c</b> and <b>-0</b>;
it is not used for data retrieval because this setting is stored in
the index file. [gff]
</dd><dt><b>-s, --sequence </b><em>INT</em></dt><dd><p>
Column of sequence name. Option
<b>-s</b>, <b>-b</b>, <b>-e</b>, <b>-S</b>, <b>-c</b> and <b>-0</b>
are all stored in the index file and thus not used in data retrieval. [1]
</dd><dt><b>-S, --skip-lines </b><em>INT</em></dt><dd><p>
Skip first INT lines in the data file. [0]
<p>
</dd></dl>
<h1 id="QUERYING_AND_OTHER_OPTIONS"><a href="#QUERYING_AND_OTHER_OPTIONS">QUERYING AND OTHER OPTIONS</a></h1>
<dl><dt><b>-h, --print-header </b></dt><dd><p>
Print also the header/meta lines.
</dd><dt><b>-H, --only-header </b></dt><dd><p>
Print only the header/meta lines.
</dd><dt><b>-l, --list-chroms </b></dt><dd><p>
List the sequence names stored in the index file.
</dd><dt><b>-r, --reheader </b><em>FILE</em></dt><dd><p>
Replace the header with the content of FILE
</dd><dt><b>-R, --regions </b><em>FILE</em></dt><dd><p>
Restrict to regions listed in the FILE. The FILE can be BED file (requires .bed, .bed.gz, .bed.bgz 
file name extension) or a TAB-delimited file with CHROM, POS, and,  optionally,
POS_TO columns, where positions are 1-based and inclusive.  When this option is in use, the input
file may not be sorted. 
</dd><dt><b>-T, --targets </b><em>FILE</em></dt><dd><p>
Similar to 
<b>-R</b>
but the entire input will be read sequentially and regions not listed in FILE will be skipped.
</dd><dt><b>-D </b></dt><dd><p>
Do not download the index file before opening it. Valid for remote files only.
</dd><dt><b>--cache </b><em>INT</em></dt><dd><p>
Set the BGZF block cache size to INT megabytes. [10]
<p>
This is of most benefit when the
<b>-R</b>
option is used, which can cause blocks to be read more than once.
Setting the size to 0 will disable the cache.
</dd><dt><b>--separate-regions</b></dt><dd><p>
This option can be used when multiple regions are supplied in the command line
and the user needs to quickly see which file records belong to which region.
For this, a line with the name of the region, preceded by the file specific
comment symbol, is inserted in the output before its corresponding group of
records.
</dd><dt><b>--verbosity </b><em>INT</em></dt><dd><p>
Set verbosity of logging messages printed to stderr.
The default is 3, which turns on error and warning messages;
2 reduces warning messages;
1 prints only error messages and 0 is mostly silent.
Values higher than 3 produce additional informational and debugging messages.
</dd></dl><p>
<h1 id="EXAMPLE"><a href="#EXAMPLE">EXAMPLE</a></h1>
(grep "^#" in.gff; grep -v "^#" in.gff | sort -t"`printf '\t'`" -k1,1 -k4,4n) | bgzip &gt; sorted.gff.gz;
<p>
tabix -p gff sorted.gff.gz;
<p>
tabix sorted.gff.gz chr1:10,000,000-20,000,000;
<p>
<h1 id="NOTES"><a href="#NOTES">NOTES</a></h1>
It is straightforward to achieve overlap queries using the standard
B-tree index (with or without binning) implemented in all SQL databases,
or the R-tree index in PostgreSQL and Oracle. But there are still many
reasons to use tabix. Firstly, tabix directly works with a lot of widely
used TAB-delimited formats such as GFF/GTF and BED. We do not need to
design database schema or specialized binary formats. Data do not need
to be duplicated in different formats, either. Secondly, tabix works on
compressed data files while most SQL databases do not. The GenCode
annotation GTF can be compressed down to 4%.  Thirdly, tabix is
fast. The same indexing algorithm is known to work efficiently for an
alignment with a few billion short reads. SQL databases probably cannot
easily handle data at this scale. Last but not the least, tabix supports
remote data retrieval. One can put the data file and the index at an FTP
or HTTP server, and other users or even web services will be able to get
a slice without downloading the entire file.
<p>
<h1 id="AUTHOR"><a href="#AUTHOR">AUTHOR</a></h1>
<p>
Tabix was written by Heng Li. The BGZF library was originally
implemented by Bob Handsaker and modified by Heng Li for remote file
access and in-memory caching.
<p>
<h1 id="SEE_ALSO"><a href="#SEE_ALSO">SEE ALSO</a></h1>
<a href="bgzip.html"><em>bgzip</em></a> (1),
<a href="samtools.html"><em>samtools</em></a> (1)
